<!DOCTYPE html>
<!-- saved from url=(0022)http://localhost:6090/ -->
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">


<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Zheng Cui">
<meta name="dcterms.date" content="2023-09-11">

<title>HW2</title>
<style class="anchorjs"></style><style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="./HW2_Zheng_files/clipboard.min.js.下载"></script>
<script src="./HW2_Zheng_files/quarto.js.下载"></script>
<script src="./HW2_Zheng_files/popper.min.js.下载"></script>
<script src="./HW2_Zheng_files/tippy.umd.min.js.下载"></script>
<script src="./HW2_Zheng_files/anchor.min.js.下载"></script>
<link href="./HW2_Zheng_files/tippy.css" rel="stylesheet">
<link href="./HW2_Zheng_files/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="./HW2_Zheng_files/bootstrap.min.js.下载"></script>
<link href="./HW2_Zheng_files/bootstrap-icons.css" rel="stylesheet">
<link href="./HW2_Zheng_files/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


<style type="text/css" data-styled-id="rCS1w3zcxh" data-styled-count="1">.rCS1w3zcxh {
position: absolute;
width: 1px;
height: 1px;
left: 0;
top: 0;
z-index: 3000;
--moveable-color: #4af;
--zoom: 1;
--zoompx: 1px;
--moveable-line-padding: 0;
--moveable-control-padding: 0;
will-change: transform;
outline: 1px solid transparent;
}
.rCS1w3zcxh .moveable-control-box {
z-index: 0;
}
.rCS1w3zcxh .moveable-line, .rCS1w3zcxh .moveable-control {
position: absolute;
left: 0;
top: 0;
will-change: transform;
}
.rCS1w3zcxh .moveable-control {
width: 14px;
height: 14px;
border-radius: 50%;
border: 2px solid #fff;
box-sizing: border-box;
background: #4af;
background: var(--moveable-color);
margin-top: -7px;
margin-left: -7px;
border: 2px solid #fff;
z-index: 10;
}
.rCS1w3zcxh .moveable-around-control {
position: absolute;
will-change: transform;
width: calc(var(--moveable-control-padding, 20) * 1px);
height: calc(var(--moveable-control-padding, 20) * 1px);
left: calc(var(--moveable-control-padding, 20) * -0.5px);
top: calc(var(--moveable-control-padding, 20) * -0.5px);
box-sizing: border-box;
background: transparent;
z-index: 8;
cursor: alias;
transform-origin: center center;
}
.rCS1w3zcxh .moveable-around-control[data-direction*="n"] {
        top: calc(var(--moveable-control-padding, 20) * -1px);
        transform-origin: center bottom;
    }
.rCS1w3zcxh .moveable-around-control[data-direction*="w"] {
        left: calc(var(--moveable-control-padding, 20) * -1px);
        transform-origin: right center;
    }
.rCS1w3zcxh .moveable-around-control[data-direction*="s"] {
        top: 0px;
        transform-origin: center top;
    }
.rCS1w3zcxh .moveable-around-control[data-direction*="e"] {
        left: 0px;
        transform-origin: left center;
    }
.rCS1w3zcxh .moveable-around-control[data-direction*="nw"] {
        left: calc(var(--moveable-control-padding, 20) * -1px);top: calc(var(--moveable-control-padding, 20) * -1px);
        transform-origin: right bottom;
    }
.rCS1w3zcxh .moveable-around-control[data-direction*="ne"] {
        left: 0px;top: calc(var(--moveable-control-padding, 20) * -1px);
        transform-origin: left bottom;
    }
.rCS1w3zcxh .moveable-around-control[data-direction*="sw"] {
        left: calc(var(--moveable-control-padding, 20) * -1px);top: 0px;
        transform-origin: right top;
    }
.rCS1w3zcxh .moveable-around-control[data-direction*="se"] {
        left: 0px;top: 0px;
        transform-origin: left top;
    }
.rCS1w3zcxh .moveable-padding {
position: absolute;
top: 0px;
left: 0px;
width: 100px;
height: 100px;
transform-origin: 0 0;
}
.rCS1w3zcxh .moveable-line {
width: 1px;
height: 1px;
background: #4af;
background: var(--moveable-color);
transform-origin: 0px 50%;
}
.rCS1w3zcxh .moveable-line.moveable-edge {
z-index: 1;
background: transparent;
}
.rCS1w3zcxh .moveable-line.moveable-dashed {
box-sizing: border-box;
background: transparent;
}
.rCS1w3zcxh .moveable-line.moveable-dashed.moveable-horizontal {
border-top: 1px dashed #4af;
border-top-color: #4af;
border-top-color: var(--moveable-color);
}
.rCS1w3zcxh .moveable-line.moveable-dashed.moveable-vertical {
border-left: 1px dashed #4af;
border-left-color: #4af;
border-left-color: var(--moveable-color);
}
.rCS1w3zcxh .moveable-line.moveable-vertical {
transform: translateX(-50%);
}
.rCS1w3zcxh .moveable-line.moveable-horizontal {
transform: translateY(-50%);
}
.rCS1w3zcxh .moveable-line.moveable-vertical.moveable-bold {
width: 2px;
}
.rCS1w3zcxh .moveable-line.moveable-horizontal.moveable-bold {
height: 2px;
}

.rCS1w3zcxh .moveable-control.moveable-origin {
border-color: #f55;
background: #fff;
width: 12px;
height: 12px;
margin-top: -6px;
margin-left: -6px;
pointer-events: none;
}

.rCS1w3zcxh .moveable-direction[data-rotation="0"],  .moveable-view-control-rotation0 {
cursor:ns-resize;cursor: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="32px" height="32px" viewBox="0 0 32 32" ><path d="M 16,5 L 12,10 L 14.5,10 L 14.5,22 L 12,22 L 16,27 L 20,22 L 17.5,22 L 17.5,10 L 20, 10 L 16,5 Z" stroke-linejoin="round" stroke-width="1.2" fill="black" stroke="white" style="transform:rotate(0deg);transform-origin: 16px 16px"></path></svg>') 16 16, ns-resize;
}


.rCS1w3zcxh .moveable-direction[data-rotation="15"],  .moveable-view-control-rotation15 {
cursor:ns-resize;cursor: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="32px" height="32px" viewBox="0 0 32 32" ><path d="M 16,5 L 12,10 L 14.5,10 L 14.5,22 L 12,22 L 16,27 L 20,22 L 17.5,22 L 17.5,10 L 20, 10 L 16,5 Z" stroke-linejoin="round" stroke-width="1.2" fill="black" stroke="white" style="transform:rotate(15deg);transform-origin: 16px 16px"></path></svg>') 16 16, ns-resize;
}


.rCS1w3zcxh .moveable-direction[data-rotation="30"],  .moveable-view-control-rotation30 {
cursor:nesw-resize;cursor: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="32px" height="32px" viewBox="0 0 32 32" ><path d="M 16,5 L 12,10 L 14.5,10 L 14.5,22 L 12,22 L 16,27 L 20,22 L 17.5,22 L 17.5,10 L 20, 10 L 16,5 Z" stroke-linejoin="round" stroke-width="1.2" fill="black" stroke="white" style="transform:rotate(30deg);transform-origin: 16px 16px"></path></svg>') 16 16, nesw-resize;
}


.rCS1w3zcxh .moveable-direction[data-rotation="45"],  .moveable-view-control-rotation45 {
cursor:nesw-resize;cursor: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="32px" height="32px" viewBox="0 0 32 32" ><path d="M 16,5 L 12,10 L 14.5,10 L 14.5,22 L 12,22 L 16,27 L 20,22 L 17.5,22 L 17.5,10 L 20, 10 L 16,5 Z" stroke-linejoin="round" stroke-width="1.2" fill="black" stroke="white" style="transform:rotate(45deg);transform-origin: 16px 16px"></path></svg>') 16 16, nesw-resize;
}


.rCS1w3zcxh .moveable-direction[data-rotation="60"],  .moveable-view-control-rotation60 {
cursor:nesw-resize;cursor: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="32px" height="32px" viewBox="0 0 32 32" ><path d="M 16,5 L 12,10 L 14.5,10 L 14.5,22 L 12,22 L 16,27 L 20,22 L 17.5,22 L 17.5,10 L 20, 10 L 16,5 Z" stroke-linejoin="round" stroke-width="1.2" fill="black" stroke="white" style="transform:rotate(60deg);transform-origin: 16px 16px"></path></svg>') 16 16, nesw-resize;
}


.rCS1w3zcxh .moveable-direction[data-rotation="75"],  .moveable-view-control-rotation75 {
cursor:ew-resize;cursor: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="32px" height="32px" viewBox="0 0 32 32" ><path d="M 16,5 L 12,10 L 14.5,10 L 14.5,22 L 12,22 L 16,27 L 20,22 L 17.5,22 L 17.5,10 L 20, 10 L 16,5 Z" stroke-linejoin="round" stroke-width="1.2" fill="black" stroke="white" style="transform:rotate(75deg);transform-origin: 16px 16px"></path></svg>') 16 16, ew-resize;
}


.rCS1w3zcxh .moveable-direction[data-rotation="90"],  .moveable-view-control-rotation90 {
cursor:ew-resize;cursor: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="32px" height="32px" viewBox="0 0 32 32" ><path d="M 16,5 L 12,10 L 14.5,10 L 14.5,22 L 12,22 L 16,27 L 20,22 L 17.5,22 L 17.5,10 L 20, 10 L 16,5 Z" stroke-linejoin="round" stroke-width="1.2" fill="black" stroke="white" style="transform:rotate(90deg);transform-origin: 16px 16px"></path></svg>') 16 16, ew-resize;
}


.rCS1w3zcxh .moveable-direction[data-rotation="105"],  .moveable-view-control-rotation105 {
cursor:ew-resize;cursor: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="32px" height="32px" viewBox="0 0 32 32" ><path d="M 16,5 L 12,10 L 14.5,10 L 14.5,22 L 12,22 L 16,27 L 20,22 L 17.5,22 L 17.5,10 L 20, 10 L 16,5 Z" stroke-linejoin="round" stroke-width="1.2" fill="black" stroke="white" style="transform:rotate(105deg);transform-origin: 16px 16px"></path></svg>') 16 16, ew-resize;
}


.rCS1w3zcxh .moveable-direction[data-rotation="120"],  .moveable-view-control-rotation120 {
cursor:nwse-resize;cursor: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="32px" height="32px" viewBox="0 0 32 32" ><path d="M 16,5 L 12,10 L 14.5,10 L 14.5,22 L 12,22 L 16,27 L 20,22 L 17.5,22 L 17.5,10 L 20, 10 L 16,5 Z" stroke-linejoin="round" stroke-width="1.2" fill="black" stroke="white" style="transform:rotate(120deg);transform-origin: 16px 16px"></path></svg>') 16 16, nwse-resize;
}


.rCS1w3zcxh .moveable-direction[data-rotation="135"],  .moveable-view-control-rotation135 {
cursor:nwse-resize;cursor: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="32px" height="32px" viewBox="0 0 32 32" ><path d="M 16,5 L 12,10 L 14.5,10 L 14.5,22 L 12,22 L 16,27 L 20,22 L 17.5,22 L 17.5,10 L 20, 10 L 16,5 Z" stroke-linejoin="round" stroke-width="1.2" fill="black" stroke="white" style="transform:rotate(135deg);transform-origin: 16px 16px"></path></svg>') 16 16, nwse-resize;
}


.rCS1w3zcxh .moveable-direction[data-rotation="150"],  .moveable-view-control-rotation150 {
cursor:nwse-resize;cursor: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="32px" height="32px" viewBox="0 0 32 32" ><path d="M 16,5 L 12,10 L 14.5,10 L 14.5,22 L 12,22 L 16,27 L 20,22 L 17.5,22 L 17.5,10 L 20, 10 L 16,5 Z" stroke-linejoin="round" stroke-width="1.2" fill="black" stroke="white" style="transform:rotate(150deg);transform-origin: 16px 16px"></path></svg>') 16 16, nwse-resize;
}


.rCS1w3zcxh .moveable-direction[data-rotation="165"],  .moveable-view-control-rotation165 {
cursor:ns-resize;cursor: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="32px" height="32px" viewBox="0 0 32 32" ><path d="M 16,5 L 12,10 L 14.5,10 L 14.5,22 L 12,22 L 16,27 L 20,22 L 17.5,22 L 17.5,10 L 20, 10 L 16,5 Z" stroke-linejoin="round" stroke-width="1.2" fill="black" stroke="white" style="transform:rotate(165deg);transform-origin: 16px 16px"></path></svg>') 16 16, ns-resize;
}


.rCS1w3zcxh .moveable-line.moveable-direction:before {
content: "";
position: absolute;
width: 100%;
height: calc(var(--moveable-line-padding, 0) * 1px);
bottom: 0;
left: 0;
}
.rCS1w3zcxh .moveable-group {
z-index: -1;
}
.rCS1w3zcxh .moveable-area {
position: absolute;
}
.rCS1w3zcxh .moveable-area-pieces {
position: absolute;
top: 0;
left: 0;
display: none;
}
.rCS1w3zcxh .moveable-area.moveable-avoid, .rCS1w3zcxh .moveable-area.moveable-pass {
pointer-events: none;
}
.rCS1w3zcxh .moveable-area.moveable-avoid+.moveable-area-pieces {
display: block;
}
.rCS1w3zcxh .moveable-area-piece {
position: absolute;
}


.rCS1w3zcxh {
--bounds-color: #d66;
}
.rCS1w3zcxh .moveable-guideline {
pointer-events: none;
z-index: 2;
}
.rCS1w3zcxh .moveable-guideline.moveable-bounds {
background: #d66;
background: var(--bounds-color);
}
.rCS1w3zcxh .moveable-guideline-group {
position: absolute;
top: 0;
left: 0;
}
.rCS1w3zcxh .moveable-guideline-group .moveable-size-value {
position: absolute;
color: #f55;
font-size: 12px;
font-size: calc(12px * var(--zoom));
font-weight: bold;
}
.rCS1w3zcxh .moveable-guideline-group.moveable-horizontal .moveable-size-value {
transform-origin: 50% 100%;
transform: translateX(-50%);
left: 50%;
bottom: 5px;
bottom: calc(2px + 3px * var(--zoom));
}
.rCS1w3zcxh .moveable-guideline-group.moveable-vertical .moveable-size-value {
transform-origin: 0% 50%;
top: 50%;
transform: translateY(-50%);
left: 5px;
left: calc(2px + 3px * var(--zoom));
}
.rCS1w3zcxh .moveable-guideline.moveable-gap {
background: #f55;
}
.rCS1w3zcxh .moveable-size-value.moveable-gap {
color: #f55;
}

.rCS1w3zcxh .moveable-edge.moveable-edgeDraggable.moveable-line {
cursor: move;
}
.rCS1w3zcxh .moveable-rotation {
position: absolute;
height: 40px;
width: 1px;
transform-origin: 50% 100%;
height: calc(40px * var(--zoom));
top: auto;
left: 0;
bottom: 100%;
will-change: transform;
}
.rCS1w3zcxh .moveable-rotation .moveable-rotation-line {
display: block;
width: 100%;
height: 100%;
transform-origin: 50% 50%;
}
.rCS1w3zcxh .moveable-rotation .moveable-rotation-control {
border-color: #4af;
border-color: var(--moveable-color);
background:#fff;
cursor: alias;
}
 .moveable-view-rotation-dragging, .rCS1w3zcxh .moveable-rotatable.moveable-direction.moveable-control {
cursor: alias;
}
.rCS1w3zcxh .moveable-rotatable.moveable-direction.moveable-control.moveable-move {
cursor: move;
}

.rCS1w3zcxh[data-able-origindraggable] .moveable-control.moveable-origin {
pointer-events: auto;
}
.rCS1w3zcxh .moveable-control.moveable-clip-control {
background: #6d6;
cursor: pointer;
}
.rCS1w3zcxh .moveable-control.moveable-clip-control.moveable-clip-radius {
background: #d66;
}
.rCS1w3zcxh .moveable-line.moveable-clip-line {
background: #6e6;
cursor: move;
z-index: 1;
}
.rCS1w3zcxh .moveable-clip-area {
position: absolute;
top: 0;
left: 0;
}
.rCS1w3zcxh .moveable-clip-ellipse {
position: absolute;
cursor: move;
border: 1px solid #6d6;
border: var(--zoompx) solid #6d6;
border-radius: 50%;
transform-origin: 0px 0px;
}
.rCS1w3zcxh {
--bounds-color: #d66;
}
.rCS1w3zcxh .moveable-guideline {
pointer-events: none;
z-index: 2;
}
.rCS1w3zcxh .moveable-line.moveable-guideline.moveable-bounds {
background: #d66;
background: var(--bounds-color);
}
.rCS1w3zcxh .moveable-control.moveable-border-radius {
background: #d66;
cursor: pointer;
z-index: 3;
}
.rCS1w3zcxh .moveable-control.moveable-border-radius.moveable-vertical {
background: #d6d;
z-index: 2;
}
.rCS1w3zcxh .moveable-control.moveable-border-radius.moveable-virtual {
opacity: 0.5;
z-index: 1;
}
.rCS1w3zcxh.moveable-round-line-clickable .moveable-line.moveable-direction {
cursor: pointer;
}</style></head>

<body class="fullcontent quarto-light">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">HW2</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Zheng Cui </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">September 11, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="quarto" class="level2">
<h2 class="anchored" data-anchor-id="quarto">Quarto<a class="anchorjs-link " aria-label="Anchor" data-anchorjs-icon="" href="http://localhost:6090/#quarto" style="font: 1em / 1 anchorjs-icons; padding-left: 0.375em;"></a></h2>
<p>Quarto enables you to weave together content and executable code into a finished document. To learn more about Quarto see <a href="https://quarto.org/" class="uri">https://quarto.org</a>.</p>
</section>
<section id="running-code" class="level2">
<h2 class="anchored" data-anchor-id="running-code">Running Code<a class="anchorjs-link " aria-label="Anchor" data-anchorjs-icon="" href="http://localhost:6090/#running-code" style="font: 1em / 1 anchorjs-icons; padding-left: 0.375em;"></a></h2>
<p>When you click the <strong>Render</strong> button a document will be generated that includes both content and the output of embedded code. You can embed code like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="http://localhost:6090/#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> <span class="sc">+</span> <span class="dv">1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
</div>
<p>You can add options to executable code like this</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4</code></pre>
</div>
</div>
<p>The <code>echo: false</code> option disables the printing of code (only output is displayed).</p>
<section id="numbers-of-genes-on-each-chromosome" class="level3">
<h3 class="anchored" data-anchor-id="numbers-of-genes-on-each-chromosome">1. Numbers of genes on each chromosome<a class="anchorjs-link " aria-label="Anchor" data-anchorjs-icon="" href="http://localhost:6090/#numbers-of-genes-on-each-chromosome" style="font: 1em / 1 anchorjs-icons; padding-left: 0.375em;"></a></h3>
<p>(20%) Visualize using bar plot (function barplot()) the total number of genes on each chromosome (1-22, X, Y). Report the chromosomes with the min and max number of genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="http://localhost:6090/#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">#  import  GFF3 table</span></span>
<span id="cb4-2"><a href="http://localhost:6090/#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb4-3"><a href="http://localhost:6090/#cb4-3" aria-hidden="true" tabindex="-1"></a>file <span class="ot">&lt;-</span> <span class="st">"E:/Language/R/gencode.v44.primary_assembly.annotation.gff3"</span></span>
<span id="cb4-4"><a href="http://localhost:6090/#cb4-4" aria-hidden="true" tabindex="-1"></a>GFF <span class="ot">&lt;-</span> <span class="fu">read.table</span>(file, <span class="at">header =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb4-5"><a href="http://localhost:6090/#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(GFF) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"seqid"</span>, <span class="st">"source"</span>, <span class="st">"type"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>,</span>
<span id="cb4-6"><a href="http://localhost:6090/#cb4-6" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"score"</span>, <span class="st">"strand"</span>, <span class="st">"phase"</span>, <span class="st">"attributes"</span>)</span>
<span id="cb4-7"><a href="http://localhost:6090/#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(GFF) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(GFF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="http://localhost:6090/#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize and calculate minimum and maximum values</span></span>
<span id="cb5-2"><a href="http://localhost:6090/#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="http://localhost:6090/#cb5-3" aria-hidden="true" tabindex="-1"></a>chromosome_counts <span class="ot">&lt;-</span> <span class="fu">table</span>(GFF<span class="sc">$</span>seqid)</span>
<span id="cb5-4"><a href="http://localhost:6090/#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="http://localhost:6090/#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Select only data for chr1 to chr22, X, and Y chromosomes</span></span>
<span id="cb5-6"><a href="http://localhost:6090/#cb5-6" aria-hidden="true" tabindex="-1"></a>selected_chromosomes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr3"</span>, <span class="st">"chr4"</span>, <span class="st">"chr5"</span>, <span class="st">"chr6"</span>, <span class="st">"chr7"</span>, <span class="st">"chr8"</span>, <span class="st">"chr9"</span>, <span class="st">"chr10"</span>,</span>
<span id="cb5-7"><a href="http://localhost:6090/#cb5-7" aria-hidden="true" tabindex="-1"></a>                          <span class="st">"chr11"</span>, <span class="st">"chr12"</span>, <span class="st">"chr13"</span>, <span class="st">"chr14"</span>, <span class="st">"chr15"</span>, <span class="st">"chr16"</span>, <span class="st">"chr17"</span>, <span class="st">"chr18"</span>, <span class="st">"chr19"</span>,</span>
<span id="cb5-8"><a href="http://localhost:6090/#cb5-8" aria-hidden="true" tabindex="-1"></a>                          <span class="st">"chr20"</span>, <span class="st">"chr21"</span>, <span class="st">"chr22"</span>, <span class="st">"chrX"</span>, <span class="st">"chrY"</span>)</span>
<span id="cb5-9"><a href="http://localhost:6090/#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="http://localhost:6090/#cb5-10" aria-hidden="true" tabindex="-1"></a>chromosome_counts <span class="ot">&lt;-</span> chromosome_counts[selected_chromosomes]</span>
<span id="cb5-11"><a href="http://localhost:6090/#cb5-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-12"><a href="http://localhost:6090/#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a larger bar plot to visualize gene counts</span></span>
<span id="cb5-13"><a href="http://localhost:6090/#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width =</span> <span class="dv">10</span>, <span class="at">repr.plot.height =</span> <span class="dv">4</span>)  <span class="co"># Adjust the plot size as needed</span></span>
<span id="cb5-14"><a href="http://localhost:6090/#cb5-14" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(chromosome_counts,</span>
<span id="cb5-15"><a href="http://localhost:6090/#cb5-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">names.arg =</span> <span class="fu">names</span>(chromosome_counts),</span>
<span id="cb5-16"><a href="http://localhost:6090/#cb5-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Total Number of Genes on Each Selected Chromosome"</span>,</span>
<span id="cb5-17"><a href="http://localhost:6090/#cb5-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">"Chromosome"</span>,</span>
<span id="cb5-18"><a href="http://localhost:6090/#cb5-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"Number of Genes"</span>,</span>
<span id="cb5-19"><a href="http://localhost:6090/#cb5-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="st">"blue"</span>,</span>
<span id="cb5-20"><a href="http://localhost:6090/#cb5-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.first =</span> <span class="fu">grid</span>(),</span>
<span id="cb5-21"><a href="http://localhost:6090/#cb5-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">las =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="./HW2_Zheng_files/unnamed-chunk-4-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="http://localhost:6090/#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the chromosomes with the minimum and maximum number of genes</span></span>
<span id="cb6-2"><a href="http://localhost:6090/#cb6-2" aria-hidden="true" tabindex="-1"></a>min_gene_chromosome <span class="ot">&lt;-</span> <span class="fu">names</span>(chromosome_counts)[<span class="fu">which.min</span>(chromosome_counts)]</span>
<span id="cb6-3"><a href="http://localhost:6090/#cb6-3" aria-hidden="true" tabindex="-1"></a>max_gene_chromosome <span class="ot">&lt;-</span> <span class="fu">names</span>(chromosome_counts)[<span class="fu">which.max</span>(chromosome_counts)]</span>
<span id="cb6-4"><a href="http://localhost:6090/#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="http://localhost:6090/#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Chromosome with the min number of genes:"</span>, min_gene_chromosome, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Chromosome with the min number of genes: chrY </code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="http://localhost:6090/#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Chromosome with the max number of genes:"</span>, max_gene_chromosome, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Chromosome with the max number of genes: chr1 </code></pre>
</div>
</div>
</section>
<section id="gene-density-on-each-chromosome" class="level3">
<h3 class="anchored" data-anchor-id="gene-density-on-each-chromosome">2. Gene density on each chromosome<a class="anchorjs-link " aria-label="Anchor" data-anchorjs-icon="" href="http://localhost:6090/#gene-density-on-each-chromosome" style="font: 1em / 1 anchorjs-icons; padding-left: 0.375em;"></a></h3>
<p>(20%) Visualize the density of genes on each chromosome. Density is defined by the total number of genes divided by the length of the chromosome. You can use the maximum end coordinate value to approximate the length of a chromosome. Point out the chromosomes with the minimum and maximum gene density, respectively.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="http://localhost:6090/#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a subset of GFF data for chr1-22, X, and Y chromosomes</span></span>
<span id="cb10-2"><a href="http://localhost:6090/#cb10-2" aria-hidden="true" tabindex="-1"></a>selected_chromosomes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr3"</span>, <span class="st">"chr4"</span>, <span class="st">"chr5"</span>, <span class="st">"chr6"</span>, <span class="st">"chr7"</span>, <span class="st">"chr8"</span>, <span class="st">"chr9"</span>, <span class="st">"chr10"</span>,</span>
<span id="cb10-3"><a href="http://localhost:6090/#cb10-3" aria-hidden="true" tabindex="-1"></a>                          <span class="st">"chr11"</span>, <span class="st">"chr12"</span>, <span class="st">"chr13"</span>, <span class="st">"chr14"</span>, <span class="st">"chr15"</span>, <span class="st">"chr16"</span>, <span class="st">"chr17"</span>, <span class="st">"chr18"</span>, <span class="st">"chr19"</span>,</span>
<span id="cb10-4"><a href="http://localhost:6090/#cb10-4" aria-hidden="true" tabindex="-1"></a>                          <span class="st">"chr20"</span>, <span class="st">"chr21"</span>, <span class="st">"chr22"</span>, <span class="st">"chrX"</span>, <span class="st">"chrY"</span>)</span>
<span id="cb10-5"><a href="http://localhost:6090/#cb10-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-6"><a href="http://localhost:6090/#cb10-6" aria-hidden="true" tabindex="-1"></a>selected_GFF <span class="ot">&lt;-</span> GFF[GFF<span class="sc">$</span>seqid <span class="sc">%in%</span> selected_chromosomes, ]</span>
<span id="cb10-7"><a href="http://localhost:6090/#cb10-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-8"><a href="http://localhost:6090/#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate gene density and Visualize and calculate min and max values</span></span>
<span id="cb10-9"><a href="http://localhost:6090/#cb10-9" aria-hidden="true" tabindex="-1"></a>chromosome_lengths <span class="ot">&lt;-</span> <span class="fu">aggregate</span>(selected_GFF<span class="sc">$</span>end, <span class="fu">list</span>(<span class="at">seqid =</span> selected_GFF<span class="sc">$</span>seqid), max)</span>
<span id="cb10-10"><a href="http://localhost:6090/#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(chromosome_lengths) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"seqid"</span>, <span class="st">"chromosome_length"</span>)</span>
<span id="cb10-11"><a href="http://localhost:6090/#cb10-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-12"><a href="http://localhost:6090/#cb10-12" aria-hidden="true" tabindex="-1"></a>chromosome_gene_counts <span class="ot">&lt;-</span> <span class="fu">aggregate</span>(selected_GFF<span class="sc">$</span>seqid, <span class="fu">list</span>(<span class="at">seqid =</span> selected_GFF<span class="sc">$</span>seqid), length)</span>
<span id="cb10-13"><a href="http://localhost:6090/#cb10-13" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(chromosome_gene_counts) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"seqid"</span>, <span class="st">"num_genes"</span>)</span>
<span id="cb10-14"><a href="http://localhost:6090/#cb10-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-15"><a href="http://localhost:6090/#cb10-15" aria-hidden="true" tabindex="-1"></a>chromosome_gene_density <span class="ot">&lt;-</span> <span class="fu">merge</span>(chromosome_lengths, chromosome_gene_counts, <span class="at">by =</span> <span class="st">"seqid"</span>)</span>
<span id="cb10-16"><a href="http://localhost:6090/#cb10-16" aria-hidden="true" tabindex="-1"></a>chromosome_gene_density<span class="sc">$</span>gene_density <span class="ot">&lt;-</span> chromosome_gene_density<span class="sc">$</span>num_genes <span class="sc">/</span> chromosome_gene_density<span class="sc">$</span>chromosome_length</span>
<span id="cb10-17"><a href="http://localhost:6090/#cb10-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-18"><a href="http://localhost:6090/#cb10-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a bar plot to visualize gene density</span></span>
<span id="cb10-19"><a href="http://localhost:6090/#cb10-19" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(chromosome_gene_density<span class="sc">$</span>gene_density, </span>
<span id="cb10-20"><a href="http://localhost:6090/#cb10-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">names.arg =</span> chromosome_gene_density<span class="sc">$</span>seqid, </span>
<span id="cb10-21"><a href="http://localhost:6090/#cb10-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="fu">nrow</span>(chromosome_gene_density)),  <span class="co"># Assign colors</span></span>
<span id="cb10-22"><a href="http://localhost:6090/#cb10-22" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">"Chromosome"</span>, </span>
<span id="cb10-23"><a href="http://localhost:6090/#cb10-23" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"Gene Density"</span>, </span>
<span id="cb10-24"><a href="http://localhost:6090/#cb10-24" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Gene Density on Each Selected Chromosome"</span>,</span>
<span id="cb10-25"><a href="http://localhost:6090/#cb10-25" aria-hidden="true" tabindex="-1"></a>        <span class="at">las =</span> <span class="dv">2</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="./HW2_Zheng_files/unnamed-chunk-5-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="http://localhost:6090/#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify chromosomes with min and max gene density</span></span>
<span id="cb11-2"><a href="http://localhost:6090/#cb11-2" aria-hidden="true" tabindex="-1"></a>min_density_chromosome <span class="ot">&lt;-</span> chromosome_gene_density<span class="sc">$</span>seqid[<span class="fu">which.min</span>(chromosome_gene_density<span class="sc">$</span>gene_density)]</span>
<span id="cb11-3"><a href="http://localhost:6090/#cb11-3" aria-hidden="true" tabindex="-1"></a>max_density_chromosome <span class="ot">&lt;-</span> chromosome_gene_density<span class="sc">$</span>seqid[<span class="fu">which.max</span>(chromosome_gene_density<span class="sc">$</span>gene_density)]</span>
<span id="cb11-4"><a href="http://localhost:6090/#cb11-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-5"><a href="http://localhost:6090/#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Chromosome with the min gene density:"</span>, min_density_chromosome, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Chromosome with the min gene density: chrY </code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="http://localhost:6090/#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Chromosome with the max gene density:"</span>, max_density_chromosome, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Chromosome with the max gene density: chr19 </code></pre>
</div>
</div>
</section>
<section id="numbers-and-density-of-transcripts-on-each-chromosome" class="level3">
<h3 class="anchored" data-anchor-id="numbers-and-density-of-transcripts-on-each-chromosome">3. Numbers and density of transcripts on each chromosome<a class="anchorjs-link " aria-label="Anchor" data-anchorjs-icon="" href="http://localhost:6090/#numbers-and-density-of-transcripts-on-each-chromosome" style="font: 1em / 1 anchorjs-icons; padding-left: 0.375em;"></a></h3>
<p>(20%) Repeat 1 and 2 for transcripts.<br>
#### 3.1 Repeat 1</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="http://localhost:6090/#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter GFF data for selected chromosomes and transcript type</span></span>
<span id="cb15-2"><a href="http://localhost:6090/#cb15-2" aria-hidden="true" tabindex="-1"></a>selected_GFF <span class="ot">&lt;-</span> GFF[GFF<span class="sc">$</span>seqid <span class="sc">%in%</span> selected_chromosomes <span class="sc">&amp;</span> GFF<span class="sc">$</span>type <span class="sc">==</span> <span class="st">"transcript"</span>, ]</span>
<span id="cb15-3"><a href="http://localhost:6090/#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="http://localhost:6090/#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the number of transcripts for each chromosome</span></span>
<span id="cb15-5"><a href="http://localhost:6090/#cb15-5" aria-hidden="true" tabindex="-1"></a>chromosome_transcript_counts <span class="ot">&lt;-</span> <span class="fu">table</span>(selected_GFF<span class="sc">$</span>seqid)</span>
<span id="cb15-6"><a href="http://localhost:6090/#cb15-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-7"><a href="http://localhost:6090/#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a larger bar plot to visualize transcript counts</span></span>
<span id="cb15-8"><a href="http://localhost:6090/#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width =</span> <span class="dv">10</span>, <span class="at">repr.plot.height =</span> <span class="dv">4</span>)  <span class="co"># Adjust the plot size as needed</span></span>
<span id="cb15-9"><a href="http://localhost:6090/#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(chromosome_transcript_counts, </span>
<span id="cb15-10"><a href="http://localhost:6090/#cb15-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">names.arg =</span> <span class="fu">names</span>(chromosome_transcript_counts), </span>
<span id="cb15-11"><a href="http://localhost:6090/#cb15-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Total Number of Transcripts on Each Chromosome"</span>,</span>
<span id="cb15-12"><a href="http://localhost:6090/#cb15-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">"Chromosome"</span>,</span>
<span id="cb15-13"><a href="http://localhost:6090/#cb15-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"Transcript Counts"</span>,</span>
<span id="cb15-14"><a href="http://localhost:6090/#cb15-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="st">"blue"</span>,</span>
<span id="cb15-15"><a href="http://localhost:6090/#cb15-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.first=</span><span class="fu">grid</span>(),</span>
<span id="cb15-16"><a href="http://localhost:6090/#cb15-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">las =</span> <span class="fl">3.5</span>)   </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="./HW2_Zheng_files/unnamed-chunk-6-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="http://localhost:6090/#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the minimum and maximum</span></span>
<span id="cb16-2"><a href="http://localhost:6090/#cb16-2" aria-hidden="true" tabindex="-1"></a>min_transcript_chromosome <span class="ot">&lt;-</span> <span class="fu">names</span>(chromosome_transcript_counts)[<span class="fu">which.min</span>(chromosome_transcript_counts)]</span>
<span id="cb16-3"><a href="http://localhost:6090/#cb16-3" aria-hidden="true" tabindex="-1"></a>max_transcript_chromosome <span class="ot">&lt;-</span> <span class="fu">names</span>(chromosome_transcript_counts)[<span class="fu">which.max</span>(chromosome_transcript_counts)]</span>
<span id="cb16-4"><a href="http://localhost:6090/#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="http://localhost:6090/#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Chromosome with the minimum transcript counts:"</span>, min_transcript_chromosome, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Chromosome with the minimum transcript counts: chrY </code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="http://localhost:6090/#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Chromosome with the maximum transcript counts:"</span>, max_transcript_chromosome, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Chromosome with the maximum transcript counts: chr1 </code></pre>
</div>
</div>
<section id="repeat-2" class="level4">
<h4 class="anchored" data-anchor-id="repeat-2">3.2 Repeat 2<a class="anchorjs-link " aria-label="Anchor" data-anchorjs-icon="" href="http://localhost:6090/#repeat-2" style="font: 1em / 1 anchorjs-icons; padding-left: 0.375em;"></a></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="http://localhost:6090/#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate transcript density and visualize, and calculate min and max values</span></span>
<span id="cb20-2"><a href="http://localhost:6090/#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter GFF to include only chr1-22 and X,Y chromosomes</span></span>
<span id="cb20-3"><a href="http://localhost:6090/#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="http://localhost:6090/#cb20-4" aria-hidden="true" tabindex="-1"></a>chromosomes_of_interest <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr3"</span>, <span class="st">"chr4"</span>, <span class="st">"chr5"</span>, <span class="st">"chr6"</span>, <span class="st">"chr7"</span>, <span class="st">"chr8"</span>, <span class="st">"chr9"</span>, <span class="st">"chr10"</span>, <span class="st">"chr11"</span>, <span class="st">"chr12"</span>, <span class="st">"chr13"</span>, <span class="st">"chr14"</span>, <span class="st">"chr15"</span>, <span class="st">"chr16"</span>, <span class="st">"chr17"</span>, <span class="st">"chr18"</span>, <span class="st">"chr19"</span>, <span class="st">"chr20"</span>, <span class="st">"chr21"</span>, <span class="st">"chr22"</span>, <span class="st">"chrX"</span>, <span class="st">"chrY"</span>)</span>
<span id="cb20-5"><a href="http://localhost:6090/#cb20-5" aria-hidden="true" tabindex="-1"></a>GFF_filtered <span class="ot">&lt;-</span> GFF[GFF<span class="sc">$</span>seqid <span class="sc">%in%</span> chromosomes_of_interest, ]</span>
<span id="cb20-6"><a href="http://localhost:6090/#cb20-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-7"><a href="http://localhost:6090/#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate transcript density and Visualize, and calculate min and max values</span></span>
<span id="cb20-8"><a href="http://localhost:6090/#cb20-8" aria-hidden="true" tabindex="-1"></a>chromosome_lengths <span class="ot">&lt;-</span> <span class="fu">aggregate</span>(GFF_filtered<span class="sc">$</span>end[GFF_filtered<span class="sc">$</span>type <span class="sc">==</span> <span class="st">"transcript"</span>], <span class="fu">list</span>(<span class="at">seqid =</span> GFF_filtered<span class="sc">$</span>seqid[GFF_filtered<span class="sc">$</span>type <span class="sc">==</span> <span class="st">"transcript"</span>]), max)</span>
<span id="cb20-9"><a href="http://localhost:6090/#cb20-9" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(chromosome_lengths) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"seqid"</span>, <span class="st">"chromosome_length"</span>)</span>
<span id="cb20-10"><a href="http://localhost:6090/#cb20-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-11"><a href="http://localhost:6090/#cb20-11" aria-hidden="true" tabindex="-1"></a>chromosome_transcript_counts <span class="ot">&lt;-</span> <span class="fu">aggregate</span>(GFF_filtered<span class="sc">$</span>seqid[GFF_filtered<span class="sc">$</span>type <span class="sc">==</span> <span class="st">"transcript"</span>], <span class="fu">list</span>(<span class="at">seqid =</span> GFF_filtered<span class="sc">$</span>seqid[GFF_filtered<span class="sc">$</span>type <span class="sc">==</span> <span class="st">"transcript"</span>]), length)</span>
<span id="cb20-12"><a href="http://localhost:6090/#cb20-12" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(chromosome_transcript_counts) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"seqid"</span>, <span class="st">"num_transcripts"</span>)</span>
<span id="cb20-13"><a href="http://localhost:6090/#cb20-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-14"><a href="http://localhost:6090/#cb20-14" aria-hidden="true" tabindex="-1"></a>chromosome_transcript_density <span class="ot">&lt;-</span> <span class="fu">merge</span>(chromosome_lengths, chromosome_transcript_counts, <span class="at">by =</span> <span class="st">"seqid"</span>)</span>
<span id="cb20-15"><a href="http://localhost:6090/#cb20-15" aria-hidden="true" tabindex="-1"></a>chromosome_transcript_density<span class="sc">$</span>transcript_density <span class="ot">&lt;-</span> chromosome_transcript_density<span class="sc">$</span>num_transcripts <span class="sc">/</span> chromosome_transcript_density<span class="sc">$</span>chromosome_length</span>
<span id="cb20-16"><a href="http://localhost:6090/#cb20-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-17"><a href="http://localhost:6090/#cb20-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a bar plot to visualize transcript density with rotated labels</span></span>
<span id="cb20-18"><a href="http://localhost:6090/#cb20-18" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(chromosome_transcript_density<span class="sc">$</span>transcript_density, </span>
<span id="cb20-19"><a href="http://localhost:6090/#cb20-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">names.arg =</span> chromosome_transcript_density<span class="sc">$</span>seqid, </span>
<span id="cb20-20"><a href="http://localhost:6090/#cb20-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="fu">nrow</span>(chromosome_transcript_density)),  <span class="co"># Assign colors</span></span>
<span id="cb20-21"><a href="http://localhost:6090/#cb20-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">"Chromosome"</span>, </span>
<span id="cb20-22"><a href="http://localhost:6090/#cb20-22" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"Transcript Density"</span>, </span>
<span id="cb20-23"><a href="http://localhost:6090/#cb20-23" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Transcript Density on Each Chromosome"</span>,</span>
<span id="cb20-24"><a href="http://localhost:6090/#cb20-24" aria-hidden="true" tabindex="-1"></a>        <span class="at">las =</span> <span class="dv">2</span>)  <span class="co"># Rotate x-axis labels 90 degrees</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="./HW2_Zheng_files/unnamed-chunk-7-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="http://localhost:6090/#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify chromosomes with min and max transcript density</span></span>
<span id="cb21-2"><a href="http://localhost:6090/#cb21-2" aria-hidden="true" tabindex="-1"></a>min_transcript_density_chromosome <span class="ot">&lt;-</span> chromosome_transcript_density<span class="sc">$</span>seqid[<span class="fu">which.min</span>(chromosome_transcript_density<span class="sc">$</span>transcript_density)]</span>
<span id="cb21-3"><a href="http://localhost:6090/#cb21-3" aria-hidden="true" tabindex="-1"></a>max_transcript_density_chromosome <span class="ot">&lt;-</span> chromosome_transcript_density<span class="sc">$</span>seqid[<span class="fu">which.max</span>(chromosome_transcript_density<span class="sc">$</span>transcript_density)]</span>
<span id="cb21-4"><a href="http://localhost:6090/#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="http://localhost:6090/#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Chromosome with the min transcript density:"</span>, min_transcript_density_chromosome, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Chromosome with the min transcript density: chrY </code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="http://localhost:6090/#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Chromosome with the max transcript density:"</span>, max_transcript_density_chromosome, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Chromosome with the max transcript density: chr19 </code></pre>
</div>
</div>
</section>
</section>
<section id="numbers-and-density-of-cdss-on-each-chromosome" class="level3">
<h3 class="anchored" data-anchor-id="numbers-and-density-of-cdss-on-each-chromosome">4. Numbers and density of CDSs on each chromosome<a class="anchorjs-link " aria-label="Anchor" data-anchorjs-icon="" href="http://localhost:6090/#numbers-and-density-of-cdss-on-each-chromosome" style="font: 1em / 1 anchorjs-icons; padding-left: 0.375em;"></a></h3>
<p>(20%) Repeat 1 and 2 for CDS.</p>
<section id="repeat-1" class="level4">
<h4 class="anchored" data-anchor-id="repeat-1">4.1 Repeat 1<a class="anchorjs-link " aria-label="Anchor" data-anchorjs-icon="" href="http://localhost:6090/#repeat-1" style="font: 1em / 1 anchorjs-icons; padding-left: 0.375em;"></a></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="http://localhost:6090/#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter GFF to include only chr1-22 and X,Y chromosomes</span></span>
<span id="cb25-2"><a href="http://localhost:6090/#cb25-2" aria-hidden="true" tabindex="-1"></a>chromosomes_of_interest <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr3"</span>, <span class="st">"chr4"</span>, <span class="st">"chr5"</span>, <span class="st">"chr6"</span>, <span class="st">"chr7"</span>, <span class="st">"chr8"</span>, <span class="st">"chr9"</span>, <span class="st">"chr10"</span>, <span class="st">"chr11"</span>, <span class="st">"chr12"</span>, <span class="st">"chr13"</span>, <span class="st">"chr14"</span>, <span class="st">"chr15"</span>, <span class="st">"chr16"</span>, <span class="st">"chr17"</span>, <span class="st">"chr18"</span>, <span class="st">"chr19"</span>, <span class="st">"chr20"</span>, <span class="st">"chr21"</span>, <span class="st">"chr22"</span>, <span class="st">"chrX"</span>, <span class="st">"chrY"</span>)</span>
<span id="cb25-3"><a href="http://localhost:6090/#cb25-3" aria-hidden="true" tabindex="-1"></a>GFF_filtered <span class="ot">&lt;-</span> GFF[GFF<span class="sc">$</span>seqid <span class="sc">%in%</span> chromosomes_of_interest, ]</span>
<span id="cb25-4"><a href="http://localhost:6090/#cb25-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-5"><a href="http://localhost:6090/#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter the data for CDS entries</span></span>
<span id="cb25-6"><a href="http://localhost:6090/#cb25-6" aria-hidden="true" tabindex="-1"></a>cds_entries <span class="ot">&lt;-</span> GFF_filtered[GFF_filtered<span class="sc">$</span>type <span class="sc">==</span> <span class="st">"CDS"</span>, ]</span>
<span id="cb25-7"><a href="http://localhost:6090/#cb25-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-8"><a href="http://localhost:6090/#cb25-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the total number of CDS entries on each chromosome</span></span>
<span id="cb25-9"><a href="http://localhost:6090/#cb25-9" aria-hidden="true" tabindex="-1"></a>chromosome_cds_counts <span class="ot">&lt;-</span> <span class="fu">table</span>(cds_entries<span class="sc">$</span>seqid)</span>
<span id="cb25-10"><a href="http://localhost:6090/#cb25-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-11"><a href="http://localhost:6090/#cb25-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Define colors for the bars (you can customize these colors)</span></span>
<span id="cb25-12"><a href="http://localhost:6090/#cb25-12" aria-hidden="true" tabindex="-1"></a>bar_colors <span class="ot">&lt;-</span> <span class="fu">rainbow</span>(<span class="fu">length</span>(chromosome_cds_counts))</span>
<span id="cb25-13"><a href="http://localhost:6090/#cb25-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-14"><a href="http://localhost:6090/#cb25-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a bar plot to visualize CDS entry counts with colors</span></span>
<span id="cb25-15"><a href="http://localhost:6090/#cb25-15" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(chromosome_cds_counts, </span>
<span id="cb25-16"><a href="http://localhost:6090/#cb25-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">names.arg =</span> <span class="fu">names</span>(chromosome_cds_counts), </span>
<span id="cb25-17"><a href="http://localhost:6090/#cb25-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">"Chromosome"</span>, </span>
<span id="cb25-18"><a href="http://localhost:6090/#cb25-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"Number of CDS Entries"</span>, </span>
<span id="cb25-19"><a href="http://localhost:6090/#cb25-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"CDS Entry Counts on Each Chromosome"</span>,</span>
<span id="cb25-20"><a href="http://localhost:6090/#cb25-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> bar_colors,</span>
<span id="cb25-21"><a href="http://localhost:6090/#cb25-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">las =</span> <span class="dv">2</span>)  <span class="co"># Specify the colors for the bars</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="./HW2_Zheng_files/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="http://localhost:6090/#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the chromosome with the minimum and maximum CDS entry counts</span></span>
<span id="cb26-2"><a href="http://localhost:6090/#cb26-2" aria-hidden="true" tabindex="-1"></a>min_cds_chromosome <span class="ot">&lt;-</span> <span class="fu">names</span>(chromosome_cds_counts)[<span class="fu">which.min</span>(chromosome_cds_counts)]</span>
<span id="cb26-3"><a href="http://localhost:6090/#cb26-3" aria-hidden="true" tabindex="-1"></a>max_cds_chromosome <span class="ot">&lt;-</span> <span class="fu">names</span>(chromosome_cds_counts)[<span class="fu">which.max</span>(chromosome_cds_counts)]</span>
<span id="cb26-4"><a href="http://localhost:6090/#cb26-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-5"><a href="http://localhost:6090/#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Chromosome with the minimum CDS entry count:"</span>, min_cds_chromosome, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Chromosome with the minimum CDS entry count: chrY </code></pre>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="http://localhost:6090/#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Chromosome with the maximum CDS entry count:"</span>, max_cds_chromosome, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Chromosome with the maximum CDS entry count: chr1 </code></pre>
</div>
</div>
</section>
<section id="repeat-2-1" class="level4">
<h4 class="anchored" data-anchor-id="repeat-2-1">4.2 Repeat 2<a class="anchorjs-link " aria-label="Anchor" data-anchorjs-icon="" href="http://localhost:6090/#repeat-2-1" style="font: 1em / 1 anchorjs-icons; padding-left: 0.375em;"></a></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="http://localhost:6090/#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter GFF to include only chr1-22 and X,Y chromosomes</span></span>
<span id="cb30-2"><a href="http://localhost:6090/#cb30-2" aria-hidden="true" tabindex="-1"></a>chromosomes_of_interest <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>, <span class="st">"chr3"</span>, <span class="st">"chr4"</span>, <span class="st">"chr5"</span>, <span class="st">"chr6"</span>, <span class="st">"chr7"</span>, <span class="st">"chr8"</span>, <span class="st">"chr9"</span>, <span class="st">"chr10"</span>, <span class="st">"chr11"</span>, <span class="st">"chr12"</span>, <span class="st">"chr13"</span>, <span class="st">"chr14"</span>, <span class="st">"chr15"</span>, <span class="st">"chr16"</span>, <span class="st">"chr17"</span>, <span class="st">"chr18"</span>, <span class="st">"chr19"</span>, <span class="st">"chr20"</span>, <span class="st">"chr21"</span>, <span class="st">"chr22"</span>, <span class="st">"chrX"</span>, <span class="st">"chrY"</span>)</span>
<span id="cb30-3"><a href="http://localhost:6090/#cb30-3" aria-hidden="true" tabindex="-1"></a>GFF_filtered <span class="ot">&lt;-</span> GFF[GFF<span class="sc">$</span>seqid <span class="sc">%in%</span> chromosomes_of_interest, ]</span>
<span id="cb30-4"><a href="http://localhost:6090/#cb30-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-5"><a href="http://localhost:6090/#cb30-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter the data for CDS entries</span></span>
<span id="cb30-6"><a href="http://localhost:6090/#cb30-6" aria-hidden="true" tabindex="-1"></a>cds_entries <span class="ot">&lt;-</span> GFF_filtered[GFF_filtered<span class="sc">$</span>type <span class="sc">==</span> <span class="st">"CDS"</span>, ]</span>
<span id="cb30-7"><a href="http://localhost:6090/#cb30-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-8"><a href="http://localhost:6090/#cb30-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the total number of CDS entries on each chromosome</span></span>
<span id="cb30-9"><a href="http://localhost:6090/#cb30-9" aria-hidden="true" tabindex="-1"></a>chromosome_cds_counts <span class="ot">&lt;-</span> <span class="fu">table</span>(cds_entries<span class="sc">$</span>seqid)</span>
<span id="cb30-10"><a href="http://localhost:6090/#cb30-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-11"><a href="http://localhost:6090/#cb30-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate chromosome lengths</span></span>
<span id="cb30-12"><a href="http://localhost:6090/#cb30-12" aria-hidden="true" tabindex="-1"></a>chromosome_lengths <span class="ot">&lt;-</span> <span class="fu">aggregate</span>(cds_entries<span class="sc">$</span>end, <span class="fu">list</span>(<span class="at">seqid =</span> cds_entries<span class="sc">$</span>seqid), max)</span>
<span id="cb30-13"><a href="http://localhost:6090/#cb30-13" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(chromosome_lengths) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"seqid"</span>, <span class="st">"chromosome_length"</span>)</span>
<span id="cb30-14"><a href="http://localhost:6090/#cb30-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-15"><a href="http://localhost:6090/#cb30-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate CDS entry density (CDS entries per base pair)</span></span>
<span id="cb30-16"><a href="http://localhost:6090/#cb30-16" aria-hidden="true" tabindex="-1"></a>chromosome_cds_density <span class="ot">&lt;-</span> <span class="fu">merge</span>(chromosome_lengths, <span class="fu">data.frame</span>(<span class="at">seqid =</span> <span class="fu">names</span>(chromosome_cds_counts), <span class="at">num_cds =</span> <span class="fu">as.vector</span>(chromosome_cds_counts)), <span class="at">by =</span> <span class="st">"seqid"</span>)</span>
<span id="cb30-17"><a href="http://localhost:6090/#cb30-17" aria-hidden="true" tabindex="-1"></a>chromosome_cds_density<span class="sc">$</span>cds_density <span class="ot">&lt;-</span> chromosome_cds_density<span class="sc">$</span>num_cds <span class="sc">/</span> chromosome_cds_density<span class="sc">$</span>chromosome_length</span>
<span id="cb30-18"><a href="http://localhost:6090/#cb30-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-19"><a href="http://localhost:6090/#cb30-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Define colors for the bars (you can customize these colors)</span></span>
<span id="cb30-20"><a href="http://localhost:6090/#cb30-20" aria-hidden="true" tabindex="-1"></a>bar_colors <span class="ot">&lt;-</span> <span class="fu">rainbow</span>(<span class="fu">length</span>(chromosome_cds_density<span class="sc">$</span>seqid))</span>
<span id="cb30-21"><a href="http://localhost:6090/#cb30-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-22"><a href="http://localhost:6090/#cb30-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a bar plot to visualize CDS entry density with colors</span></span>
<span id="cb30-23"><a href="http://localhost:6090/#cb30-23" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(chromosome_cds_density<span class="sc">$</span>cds_density, </span>
<span id="cb30-24"><a href="http://localhost:6090/#cb30-24" aria-hidden="true" tabindex="-1"></a>        <span class="at">names.arg =</span> chromosome_cds_density<span class="sc">$</span>seqid, </span>
<span id="cb30-25"><a href="http://localhost:6090/#cb30-25" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">"Chromosome"</span>, </span>
<span id="cb30-26"><a href="http://localhost:6090/#cb30-26" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"CDS Entry Density (CDS entries per bp)"</span>, </span>
<span id="cb30-27"><a href="http://localhost:6090/#cb30-27" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"CDS Entry Density on Each Chromosome"</span>,</span>
<span id="cb30-28"><a href="http://localhost:6090/#cb30-28" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> bar_colors,</span>
<span id="cb30-29"><a href="http://localhost:6090/#cb30-29" aria-hidden="true" tabindex="-1"></a>        <span class="at">las =</span> <span class="dv">2</span>)  <span class="co"># Specify the colors for the bars</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="./HW2_Zheng_files/unnamed-chunk-9-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="http://localhost:6090/#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the chromosome with the minimum and maximum CDS entry density</span></span>
<span id="cb31-2"><a href="http://localhost:6090/#cb31-2" aria-hidden="true" tabindex="-1"></a>min_cds_density_chromosome <span class="ot">&lt;-</span> chromosome_cds_density<span class="sc">$</span>seqid[<span class="fu">which.min</span>(chromosome_cds_density<span class="sc">$</span>cds_density)]</span>
<span id="cb31-3"><a href="http://localhost:6090/#cb31-3" aria-hidden="true" tabindex="-1"></a>max_cds_density_chromosome <span class="ot">&lt;-</span> chromosome_cds_density<span class="sc">$</span>seqid[<span class="fu">which.max</span>(chromosome_cds_density<span class="sc">$</span>cds_density)]</span>
<span id="cb31-4"><a href="http://localhost:6090/#cb31-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-5"><a href="http://localhost:6090/#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Chromosome with the minimum CDS entry density:"</span>, min_cds_density_chromosome, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Chromosome with the minimum CDS entry density: chrY </code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="http://localhost:6090/#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Chromosome with the maximum CDS entry density:"</span>, max_cds_density_chromosome, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Chromosome with the maximum CDS entry density: chr19 </code></pre>
</div>
</div>
</section>
</section>
<section id="overall-statistics" class="level3">
<h3 class="anchored" data-anchor-id="overall-statistics">5. Overall statistics<a class="anchorjs-link " aria-label="Anchor" data-anchorjs-icon="" href="http://localhost:6090/#overall-statistics" style="font: 1em / 1 anchorjs-icons; padding-left: 0.375em;"></a></h3>
<section id="how-many-intervals-are-annotated-in-this-version" class="level4">
<h4 class="anchored" data-anchor-id="how-many-intervals-are-annotated-in-this-version">(5%) 5.1 How many intervals are annotated in this version?<a class="anchorjs-link " aria-label="Anchor" data-anchorjs-icon="" href="http://localhost:6090/#how-many-intervals-are-annotated-in-this-version" style="font: 1em / 1 anchorjs-icons; padding-left: 0.375em;"></a></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="http://localhost:6090/#cb35-1" aria-hidden="true" tabindex="-1"></a>total_intervals <span class="ot">&lt;-</span> <span class="fu">nrow</span>(GFF)</span>
<span id="cb35-2"><a href="http://localhost:6090/#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Total number of annotated intervals:"</span>, total_intervals, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Total number of annotated intervals: 3422649 </code></pre>
</div>
</div>
</section>
<section id="what-is-the-total-number-of-genes-annotated-in-the-human-genome" class="level4">
<h4 class="anchored" data-anchor-id="what-is-the-total-number-of-genes-annotated-in-the-human-genome">(5%) 5.2 What is the total number of genes annotated in the human genome?<a class="anchorjs-link " aria-label="Anchor" data-anchorjs-icon="" href="http://localhost:6090/#what-is-the-total-number-of-genes-annotated-in-the-human-genome" style="font: 1em / 1 anchorjs-icons; padding-left: 0.375em;"></a></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="http://localhost:6090/#cb37-1" aria-hidden="true" tabindex="-1"></a>total_genes <span class="ot">&lt;-</span> <span class="fu">sum</span>(GFF<span class="sc">$</span>type <span class="sc">==</span> <span class="st">"gene"</span>)</span>
<span id="cb37-2"><a href="http://localhost:6090/#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Total number of genes in the human genome:"</span>, total_genes, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Total number of genes in the human genome: 62754 </code></pre>
</div>
</div>
</section>
<section id="how-many-are-protein-coding-genes" class="level4">
<h4 class="anchored" data-anchor-id="how-many-are-protein-coding-genes">(5%) 5.3 How many are protein coding genes?<a class="anchorjs-link " aria-label="Anchor" data-anchorjs-icon="" href="http://localhost:6090/#how-many-are-protein-coding-genes" style="font: 1em / 1 anchorjs-icons; padding-left: 0.375em;"></a></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="http://localhost:6090/#cb39-1" aria-hidden="true" tabindex="-1"></a>pattern <span class="ot">&lt;-</span> <span class="st">"gene_type=([^;]*)"</span></span>
<span id="cb39-2"><a href="http://localhost:6090/#cb39-2" aria-hidden="true" tabindex="-1"></a>  m <span class="ot">&lt;-</span> <span class="fu">regexec</span>(pattern, GFF[, <span class="st">"attributes"</span>])</span>
<span id="cb39-3"><a href="http://localhost:6090/#cb39-3" aria-hidden="true" tabindex="-1"></a>  gene_type <span class="ot">&lt;-</span> <span class="fu">sapply</span>(</span>
<span id="cb39-4"><a href="http://localhost:6090/#cb39-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">regmatches</span>(GFF[, <span class="st">"attributes"</span>], m), </span>
<span id="cb39-5"><a href="http://localhost:6090/#cb39-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">function</span>(e) {<span class="fu">return</span>(e[<span class="dv">2</span>])}) </span>
<span id="cb39-6"><a href="http://localhost:6090/#cb39-6" aria-hidden="true" tabindex="-1"></a>  <span class="co">#</span></span>
<span id="cb39-7"><a href="http://localhost:6090/#cb39-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Very slow alternative to extract gene names:    </span></span>
<span id="cb39-8"><a href="http://localhost:6090/#cb39-8" aria-hidden="true" tabindex="-1"></a>  <span class="co">#   d &lt;- data.frame(regmatches(gff3[, "tag"], m))    </span></span>
<span id="cb39-9"><a href="http://localhost:6090/#cb39-9" aria-hidden="true" tabindex="-1"></a>  <span class="co">#   gene_names &lt;- d[2, ]  </span></span>
<span id="cb39-10"><a href="http://localhost:6090/#cb39-10" aria-hidden="true" tabindex="-1"></a>  GFF<span class="sc">$</span>gene_type2 <span class="ot">=</span> gene_type</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="http://localhost:6090/#cb40-1" aria-hidden="true" tabindex="-1"></a>total_protein_coding <span class="ot">&lt;-</span> <span class="fu">sum</span>(GFF<span class="sc">$</span>gene_type2 <span class="sc">==</span> <span class="st">"protein_coding"</span> <span class="sc">&amp;</span> GFF<span class="sc">$</span>type <span class="sc">==</span> <span class="st">"gene"</span>)</span>
<span id="cb40-2"><a href="http://localhost:6090/#cb40-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-3"><a href="http://localhost:6090/#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Total number of protein_coding in the human genome:"</span>, total_protein_coding, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Total number of protein_coding in the human genome: 20070 </code></pre>
</div>
</div>
</section>
<section id="how-many-are-noncoding-genes" class="level4">
<h4 class="anchored" data-anchor-id="how-many-are-noncoding-genes">(5%) 5.4 How many are noncoding genes?<a class="anchorjs-link " aria-label="Anchor" data-anchorjs-icon="" href="http://localhost:6090/#how-many-are-noncoding-genes" style="font: 1em / 1 anchorjs-icons; padding-left: 0.375em;"></a></h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="http://localhost:6090/#cb42-1" aria-hidden="true" tabindex="-1"></a>total_noncoding <span class="ot">&lt;-</span> <span class="fu">sum</span>(GFF<span class="sc">$</span>gene_type2 <span class="sc">!=</span> <span class="st">"protein_coding"</span> <span class="sc">&amp;</span> GFF<span class="sc">$</span>type <span class="sc">==</span> <span class="st">"gene"</span>)</span>
<span id="cb42-2"><a href="http://localhost:6090/#cb42-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-3"><a href="http://localhost:6090/#cb42-3" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb42-4"><a href="http://localhost:6090/#cb42-4" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Total number of noncoding in the human genome:"</span>, total_noncoding, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Total number of noncoding in the human genome: 42684 </code></pre>
</div>
</div>
</section>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




<div class="modal__dialog micromodal-slide" id="quarto-log-error-modal" aria-hidden="true">
  <div class="modal__overlay" tabindex="2" data-micromodal-close="">
    <div class="modal__container" role="dialog" aria-modal="true" aria-labelledby="quarto-log-error-modal-title">
      <header class="modal__header">
        <h2 class="modal__title" id="quarto-log-error-modal-title">
          <i class="bi bi-exclamation-circle"></i>Error
        </h2>
        <button class="modal__close" aria-label="Close modal" data-micromodal-close=""></button>
      </header>
      <div class="modal__content" id="quarto-log-error-modal-content">
        <pre tabindex="1" id="quarto-log-error-display"></pre>
      </div>
      <footer class="modal__footer">
      </footer>
    </div>
  </div>
</div>
<style type="text/css">
#quarto-log-error-modal .modal__overlay {
  z-index: 5000;
}
#quarto-log-error-modal-content {
  border: 1px solid #dee2e6;
  overflow-y: auto;
}
#quarto-log-error-display {
  min-height: 100px;
  padding: 8px;
  margin-bottom: 0;
  font-family: monospace;
}
#quarto-log-error-display:focus {
  outline: none;
}
.modal__dialog {
  font-size: 16px;
}
.modal__overlay {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: rgba(0,0,0,0.4);
  display: flex;
  justify-content: center;
  align-items: center;
}
.modal__container {
  position: fixed;
  top: 80px;
  background-color: #fff;
  padding: 15px;
  padding-top: 10px;
  min-width: 500px;
  max-width: 90%;
  max-height: 90vh;
  border-radius: 4px;
  overflow-y: auto;
  box-sizing: border-box;
}
.modal__header {
  display: flex;
  justify-content: space-between;
  align-items: center;
}
.modal__title {
  margin-top: 0;
  margin-bottom: 0;
  font-weight: 600;
  border-bottom: none;
  padding-bottom: 10px;
  font-size: 20px;
  color: rgb(191,71,65);
  line-height: 1.25;
  box-sizing: border-box;
}
.modal__title .bi {
  margin-right: 7px;
}
.modal__close {
  background: transparent;
  border: 0;
  margin-bottom: 10px;
}
.modal__header .modal__close:before { content: "\2715"; }
.modal__content {
  margin-top: 0;
  margin-bottom: 0;
}
@keyframes mmfadeIn {
    from { opacity: 0; }
      to { opacity: 1; }
}
@keyframes mmfadeOut {
    from { opacity: 1; }
      to { opacity: 0; }
}
@keyframes mmslideIn {
  from { transform: translateY(15%); }
    to { transform: translateY(0); }
}
@keyframes mmslideOut {
    from { transform: translateY(0); }
    to { transform: translateY(-10%); }
}
.micromodal-slide {
  display: none;
}
.micromodal-slide.is-open {
  display: block;
}
.micromodal-slide[aria-hidden="false"] .modal__overlay {
  animation: mmfadeIn .3s cubic-bezier(0.0, 0.0, 0.2, 1);
}
.micromodal-slide[aria-hidden="false"] .modal__container {
  animation: mmslideIn .3s cubic-bezier(0, 0, .2, 1);
}
.micromodal-slide[aria-hidden="true"] .modal__overlay {
  animation: mmfadeOut .3s cubic-bezier(0.0, 0.0, 0.2, 1);
}
.micromodal-slide[aria-hidden="true"] .modal__container {
  animation: mmslideOut .3s cubic-bezier(0, 0, .2, 1);
}
.micromodal-slide .modal__container,
.micromodal-slide .modal__overlay {
  will-change: transform;
} 
</style>
<script type="text/javascript">
  window.backupDefine = window.define; window.define = undefined;
  /*  ansi_up.js
   *  author : Dru Nelson
   *  license : MIT
   *  http://github.com/drudru/ansi_up
   */
   (function (root, factory) {
      if (typeof define === 'function' && define.amd) {
          // AMD. Register as an anonymous module.
          define(['exports'], factory);
      } else if (typeof exports === 'object' && typeof exports.nodeName !== 'string') {
          // CommonJS
          factory(exports);
      } else {
          // Browser globals
          var exp = {};
          factory(exp);
          root.AnsiUp = exp.default;
      }
  }(this, function (exports) {
  "use strict";
  var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
      if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
      return cooked;
  };
  var PacketKind;
  (function (PacketKind) {
      PacketKind[PacketKind["EOS"] = 0] = "EOS";
      PacketKind[PacketKind["Text"] = 1] = "Text";
      PacketKind[PacketKind["Incomplete"] = 2] = "Incomplete";
      PacketKind[PacketKind["ESC"] = 3] = "ESC";
      PacketKind[PacketKind["Unknown"] = 4] = "Unknown";
      PacketKind[PacketKind["SGR"] = 5] = "SGR";
      PacketKind[PacketKind["OSCURL"] = 6] = "OSCURL";
  })(PacketKind || (PacketKind = {}));
  var AnsiUp = (function () {
      function AnsiUp() {
          this.VERSION = "5.1.0";
          this.setup_palettes();
          this._use_classes = false;
          this.bold = false;
          this.italic = false;
          this.underline = false;
          this.fg = this.bg = null;
          this._buffer = '';
          this._url_whitelist = { 'http': 1, 'https': 1 };
      }
      Object.defineProperty(AnsiUp.prototype, "use_classes", {
          get: function () {
              return this._use_classes;
          },
          set: function (arg) {
              this._use_classes = arg;
          },
          enumerable: false,
          configurable: true
      });
      Object.defineProperty(AnsiUp.prototype, "url_whitelist", {
          get: function () {
              return this._url_whitelist;
          },
          set: function (arg) {
              this._url_whitelist = arg;
          },
          enumerable: false,
          configurable: true
      });
      AnsiUp.prototype.setup_palettes = function () {
          var _this = this;
          this.ansi_colors =
              [
                  [
                      { rgb: [0, 0, 0], class_name: "ansi-black" },
                      { rgb: [187, 0, 0], class_name: "ansi-red" },
                      { rgb: [0, 187, 0], class_name: "ansi-green" },
                      { rgb: [187, 187, 0], class_name: "ansi-yellow" },
                      { rgb: [0, 0, 187], class_name: "ansi-blue" },
                      { rgb: [187, 0, 187], class_name: "ansi-magenta" },
                      { rgb: [0, 187, 187], class_name: "ansi-cyan" },
                      { rgb: [255, 255, 255], class_name: "ansi-white" }
                  ],
                  [
                      { rgb: [85, 85, 85], class_name: "ansi-bright-black" },
                      { rgb: [255, 85, 85], class_name: "ansi-bright-red" },
                      { rgb: [0, 255, 0], class_name: "ansi-bright-green" },
                      { rgb: [255, 255, 85], class_name: "ansi-bright-yellow" },
                      { rgb: [85, 85, 255], class_name: "ansi-bright-blue" },
                      { rgb: [255, 85, 255], class_name: "ansi-bright-magenta" },
                      { rgb: [85, 255, 255], class_name: "ansi-bright-cyan" },
                      { rgb: [255, 255, 255], class_name: "ansi-bright-white" }
                  ]
              ];
          this.palette_256 = [];
          this.ansi_colors.forEach(function (palette) {
              palette.forEach(function (rec) {
                  _this.palette_256.push(rec);
              });
          });
          var levels = [0, 95, 135, 175, 215, 255];
          for (var r = 0; r < 6; ++r) {
              for (var g = 0; g < 6; ++g) {
                  for (var b = 0; b < 6; ++b) {
                      var col = { rgb: [levels[r], levels[g], levels[b]], class_name: 'truecolor' };
                      this.palette_256.push(col);
                  }
              }
          }
          var grey_level = 8;
          for (var i = 0; i < 24; ++i, grey_level += 10) {
              var gry = { rgb: [grey_level, grey_level, grey_level], class_name: 'truecolor' };
              this.palette_256.push(gry);
          }
      };
      AnsiUp.prototype.escape_txt_for_html = function (txt) {
          return txt.replace(/[&<>"']/gm, function (str) {
              if (str === "&")
                  return "&amp;";
              if (str === "<")
                  return "&lt;";
              if (str === ">")
                  return "&gt;";
              if (str === "\"")
                  return "&quot;";
              if (str === "'")
                  return "&#x27;";
          });
      };
      AnsiUp.prototype.append_buffer = function (txt) {
          var str = this._buffer + txt;
          this._buffer = str;
      };
      AnsiUp.prototype.get_next_packet = function () {
          var pkt = {
              kind: PacketKind.EOS,
              text: '',
              url: ''
          };
          var len = this._buffer.length;
          if (len == 0)
              return pkt;
          var pos = this._buffer.indexOf("\x1B");
          if (pos == -1) {
              pkt.kind = PacketKind.Text;
              pkt.text = this._buffer;
              this._buffer = '';
              return pkt;
          }
          if (pos > 0) {
              pkt.kind = PacketKind.Text;
              pkt.text = this._buffer.slice(0, pos);
              this._buffer = this._buffer.slice(pos);
              return pkt;
          }
          if (pos == 0) {
              if (len == 1) {
                  pkt.kind = PacketKind.Incomplete;
                  return pkt;
              }
              var next_char = this._buffer.charAt(1);
              if ((next_char != '[') && (next_char != ']')) {
                  pkt.kind = PacketKind.ESC;
                  pkt.text = this._buffer.slice(0, 1);
                  this._buffer = this._buffer.slice(1);
                  return pkt;
              }
              if (next_char == '[') {
                  if (!this._csi_regex) {
                      this._csi_regex = rgx(__makeTemplateObject(["\n                        ^                           # beginning of line\n                                                    #\n                                                    # First attempt\n                        (?:                         # legal sequence\n                          \u001B[                      # CSI\n                          ([<-?]?)              # private-mode char\n                          ([d;]*)                    # any digits or semicolons\n                          ([ -/]?               # an intermediate modifier\n                          [@-~])                # the command\n                        )\n                        |                           # alternate (second attempt)\n                        (?:                         # illegal sequence\n                          \u001B[                      # CSI\n                          [ -~]*                # anything legal\n                          ([\0-\u001F:])              # anything illegal\n                        )\n                    "], ["\n                        ^                           # beginning of line\n                                                    #\n                                                    # First attempt\n                        (?:                         # legal sequence\n                          \\x1b\\[                      # CSI\n                          ([\\x3c-\\x3f]?)              # private-mode char\n                          ([\\d;]*)                    # any digits or semicolons\n                          ([\\x20-\\x2f]?               # an intermediate modifier\n                          [\\x40-\\x7e])                # the command\n                        )\n                        |                           # alternate (second attempt)\n                        (?:                         # illegal sequence\n                          \\x1b\\[                      # CSI\n                          [\\x20-\\x7e]*                # anything legal\n                          ([\\x00-\\x1f:])              # anything illegal\n                        )\n                    "]));
                  }
                  var match = this._buffer.match(this._csi_regex);
                  if (match === null) {
                      pkt.kind = PacketKind.Incomplete;
                      return pkt;
                  }
                  if (match[4]) {
                      pkt.kind = PacketKind.ESC;
                      pkt.text = this._buffer.slice(0, 1);
                      this._buffer = this._buffer.slice(1);
                      return pkt;
                  }
                  if ((match[1] != '') || (match[3] != 'm'))
                      pkt.kind = PacketKind.Unknown;
                  else
                      pkt.kind = PacketKind.SGR;
                  pkt.text = match[2];
                  var rpos = match[0].length;
                  this._buffer = this._buffer.slice(rpos);
                  return pkt;
              }
              if (next_char == ']') {
                  if (len < 4) {
                      pkt.kind = PacketKind.Incomplete;
                      return pkt;
                  }
                  if ((this._buffer.charAt(2) != '8')
                      || (this._buffer.charAt(3) != ';')) {
                      pkt.kind = PacketKind.ESC;
                      pkt.text = this._buffer.slice(0, 1);
                      this._buffer = this._buffer.slice(1);
                      return pkt;
                  }
                  if (!this._osc_st) {
                      this._osc_st = rgxG(__makeTemplateObject(["\n                        (?:                         # legal sequence\n                          (\u001B\\)                    # ESC                           |                           # alternate\n                          (\u0007)                      # BEL (what xterm did)\n                        )\n                        |                           # alternate (second attempt)\n                        (                           # illegal sequence\n                          [\0-\u0006]                 # anything illegal\n                          |                           # alternate\n                          [\b-\u001A]                 # anything illegal\n                          |                           # alternate\n                          [\u001C-\u001F]                 # anything illegal\n                        )\n                    "], ["\n                        (?:                         # legal sequence\n                          (\\x1b\\\\)                    # ESC \\\n                          |                           # alternate\n                          (\\x07)                      # BEL (what xterm did)\n                        )\n                        |                           # alternate (second attempt)\n                        (                           # illegal sequence\n                          [\\x00-\\x06]                 # anything illegal\n                          |                           # alternate\n                          [\\x08-\\x1a]                 # anything illegal\n                          |                           # alternate\n                          [\\x1c-\\x1f]                 # anything illegal\n                        )\n                    "]));
                  }
                  this._osc_st.lastIndex = 0;
                  {
                      var match_1 = this._osc_st.exec(this._buffer);
                      if (match_1 === null) {
                          pkt.kind = PacketKind.Incomplete;
                          return pkt;
                      }
                      if (match_1[3]) {
                          pkt.kind = PacketKind.ESC;
                          pkt.text = this._buffer.slice(0, 1);
                          this._buffer = this._buffer.slice(1);
                          return pkt;
                      }
                  }
                  {
                      var match_2 = this._osc_st.exec(this._buffer);
                      if (match_2 === null) {
                          pkt.kind = PacketKind.Incomplete;
                          return pkt;
                      }
                      if (match_2[3]) {
                          pkt.kind = PacketKind.ESC;
                          pkt.text = this._buffer.slice(0, 1);
                          this._buffer = this._buffer.slice(1);
                          return pkt;
                      }
                  }
                  if (!this._osc_regex) {
                      this._osc_regex = rgx(__makeTemplateObject(["\n                        ^                           # beginning of line\n                                                    #\n                        \u001B]8;                    # OSC Hyperlink\n                        [ -:<-~]*       # params (excluding ;)\n                        ;                           # end of params\n                        ([!-~]{0,512})        # URL capture\n                        (?:                         # ST\n                          (?:\u001B\\)                  # ESC                           |                           # alternate\n                          (?:\u0007)                    # BEL (what xterm did)\n                        )\n                        ([ -~]+)              # TEXT capture\n                        \u001B]8;;                   # OSC Hyperlink End\n                        (?:                         # ST\n                          (?:\u001B\\)                  # ESC                           |                           # alternate\n                          (?:\u0007)                    # BEL (what xterm did)\n                        )\n                    "], ["\n                        ^                           # beginning of line\n                                                    #\n                        \\x1b\\]8;                    # OSC Hyperlink\n                        [\\x20-\\x3a\\x3c-\\x7e]*       # params (excluding ;)\n                        ;                           # end of params\n                        ([\\x21-\\x7e]{0,512})        # URL capture\n                        (?:                         # ST\n                          (?:\\x1b\\\\)                  # ESC \\\n                          |                           # alternate\n                          (?:\\x07)                    # BEL (what xterm did)\n                        )\n                        ([\\x20-\\x7e]+)              # TEXT capture\n                        \\x1b\\]8;;                   # OSC Hyperlink End\n                        (?:                         # ST\n                          (?:\\x1b\\\\)                  # ESC \\\n                          |                           # alternate\n                          (?:\\x07)                    # BEL (what xterm did)\n                        )\n                    "]));
                  }
                  var match = this._buffer.match(this._osc_regex);
                  if (match === null) {
                      pkt.kind = PacketKind.ESC;
                      pkt.text = this._buffer.slice(0, 1);
                      this._buffer = this._buffer.slice(1);
                      return pkt;
                  }
                  pkt.kind = PacketKind.OSCURL;
                  pkt.url = match[1];
                  pkt.text = match[2];
                  var rpos = match[0].length;
                  this._buffer = this._buffer.slice(rpos);
                  return pkt;
              }
          }
      };
      AnsiUp.prototype.ansi_to_html = function (txt) {
          this.append_buffer(txt);
          var blocks = [];
          while (true) {
              var packet = this.get_next_packet();
              if ((packet.kind == PacketKind.EOS)
                  || (packet.kind == PacketKind.Incomplete))
                  break;
              if ((packet.kind == PacketKind.ESC)
                  || (packet.kind == PacketKind.Unknown))
                  continue;
              if (packet.kind == PacketKind.Text)
                  blocks.push(this.transform_to_html(this.with_state(packet)));
              else if (packet.kind == PacketKind.SGR)
                  this.process_ansi(packet);
              else if (packet.kind == PacketKind.OSCURL)
                  blocks.push(this.process_hyperlink(packet));
          }
          return blocks.join("");
      };
      AnsiUp.prototype.with_state = function (pkt) {
          return { bold: this.bold, italic: this.italic, underline: this.underline, fg: this.fg, bg: this.bg, text: pkt.text };
      };
      AnsiUp.prototype.process_ansi = function (pkt) {
          var sgr_cmds = pkt.text.split(';');
          while (sgr_cmds.length > 0) {
              var sgr_cmd_str = sgr_cmds.shift();
              var num = parseInt(sgr_cmd_str, 10);
              if (isNaN(num) || num === 0) {
                  this.fg = this.bg = null;
                  this.bold = false;
                  this.italic = false;
                  this.underline = false;
              }
              else if (num === 1) {
                  this.bold = true;
              }
              else if (num === 3) {
                  this.italic = true;
              }
              else if (num === 4) {
                  this.underline = true;
              }
              else if (num === 22) {
                  this.bold = false;
              }
              else if (num === 23) {
                  this.italic = false;
              }
              else if (num === 24) {
                  this.underline = false;
              }
              else if (num === 39) {
                  this.fg = null;
              }
              else if (num === 49) {
                  this.bg = null;
              }
              else if ((num >= 30) && (num < 38)) {
                  this.fg = this.ansi_colors[0][(num - 30)];
              }
              else if ((num >= 40) && (num < 48)) {
                  this.bg = this.ansi_colors[0][(num - 40)];
              }
              else if ((num >= 90) && (num < 98)) {
                  this.fg = this.ansi_colors[1][(num - 90)];
              }
              else if ((num >= 100) && (num < 108)) {
                  this.bg = this.ansi_colors[1][(num - 100)];
              }
              else if (num === 38 || num === 48) {
                  if (sgr_cmds.length > 0) {
                      var is_foreground = (num === 38);
                      var mode_cmd = sgr_cmds.shift();
                      if (mode_cmd === '5' && sgr_cmds.length > 0) {
                          var palette_index = parseInt(sgr_cmds.shift(), 10);
                          if (palette_index >= 0 && palette_index <= 255) {
                              if (is_foreground)
                                  this.fg = this.palette_256[palette_index];
                              else
                                  this.bg = this.palette_256[palette_index];
                          }
                      }
                      if (mode_cmd === '2' && sgr_cmds.length > 2) {
                          var r = parseInt(sgr_cmds.shift(), 10);
                          var g = parseInt(sgr_cmds.shift(), 10);
                          var b = parseInt(sgr_cmds.shift(), 10);
                          if ((r >= 0 && r <= 255) && (g >= 0 && g <= 255) && (b >= 0 && b <= 255)) {
                              var c = { rgb: [r, g, b], class_name: 'truecolor' };
                              if (is_foreground)
                                  this.fg = c;
                              else
                                  this.bg = c;
                          }
                      }
                  }
              }
          }
      };
      AnsiUp.prototype.transform_to_html = function (fragment) {
          var txt = fragment.text;
          if (txt.length === 0)
              return txt;
          txt = this.escape_txt_for_html(txt);
          if (!fragment.bold && !fragment.italic && !fragment.underline && fragment.fg === null && fragment.bg === null)
              return txt;
          var styles = [];
          var classes = [];
          var fg = fragment.fg;
          var bg = fragment.bg;
          if (fragment.bold)
              styles.push('font-weight:bold');
          if (fragment.italic)
              styles.push('font-style:italic');
          if (fragment.underline)
              styles.push('text-decoration:underline');
          if (!this._use_classes) {
              if (fg)
                  styles.push("color:rgb(" + fg.rgb.join(',') + ")");
              if (bg)
                  styles.push("background-color:rgb(" + bg.rgb + ")");
          }
          else {
              if (fg) {
                  if (fg.class_name !== 'truecolor') {
                      classes.push(fg.class_name + "-fg");
                  }
                  else {
                      styles.push("color:rgb(" + fg.rgb.join(',') + ")");
                  }
              }
              if (bg) {
                  if (bg.class_name !== 'truecolor') {
                      classes.push(bg.class_name + "-bg");
                  }
                  else {
                      styles.push("background-color:rgb(" + bg.rgb.join(',') + ")");
                  }
              }
          }
          var class_string = '';
          var style_string = '';
          if (classes.length)
              class_string = " class=\"" + classes.join(' ') + "\"";
          if (styles.length)
              style_string = " style=\"" + styles.join(';') + "\"";
          return "<span" + style_string + class_string + ">" + txt + "</span>";
      };
      ;
      AnsiUp.prototype.process_hyperlink = function (pkt) {
          var parts = pkt.url.split(':');
          if (parts.length < 1)
              return '';
          if (!this._url_whitelist[parts[0]])
              return '';
          var result = "<a href=\"" + this.escape_txt_for_html(pkt.url) + "\">" + this.escape_txt_for_html(pkt.text) + "</a>";
          return result;
      };
      return AnsiUp;
  }());
  function rgx(tmplObj) {
      var subst = [];
      for (var _i = 1; _i < arguments.length; _i++) {
          subst[_i - 1] = arguments[_i];
      }
      var regexText = tmplObj.raw[0];
      var wsrgx = /^\s+|\s+\n|\s*#[\s\S]*?\n|\n/gm;
      var txt2 = regexText.replace(wsrgx, '');
      return new RegExp(txt2);
  }
  function rgxG(tmplObj) {
      var subst = [];
      for (var _i = 1; _i < arguments.length; _i++) {
          subst[_i - 1] = arguments[_i];
      }
      var regexText = tmplObj.raw[0];
      var wsrgx = /^\s+|\s+\n|\s*#[\s\S]*?\n|\n/gm;
      var txt2 = regexText.replace(wsrgx, '');
      return new RegExp(txt2, 'g');
  }
      Object.defineProperty(exports, "__esModule", { value: true });
      exports.default = AnsiUp;
  }));
  // micromodal: https://github.com/Ghosh/micromodal
  !function(e,t){"object"==typeof exports&&"undefined"!=typeof module?module.exports=t():"function"==typeof define&&define.amd?define(t):(e="undefined"!=typeof globalThis?globalThis:e||self).MicroModal=t()}(this,(function(){"use strict";function e(e,t){for(var o=0;o<t.length;o++){var n=t[o];n.enumerable=n.enumerable||!1,n.configurable=!0,"value"in n&&(n.writable=!0),Object.defineProperty(e,n.key,n)}}function t(e){return function(e){if(Array.isArray(e))return o(e)}(e)||function(e){if("undefined"!=typeof Symbol&&Symbol.iterator in Object(e))return Array.from(e)}(e)||function(e,t){if(!e)return;if("string"==typeof e)return o(e,t);var n=Object.prototype.toString.call(e).slice(8,-1);"Object"===n&&e.constructor&&(n=e.constructor.name);if("Map"===n||"Set"===n)return Array.from(e);if("Arguments"===n||/^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n))return o(e,t)}(e)||function(){throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.")}()}function o(e,t){(null==t||t>e.length)&&(t=e.length);for(var o=0,n=new Array(t);o<t;o++)n[o]=e[o];return n}var n,i,a,r,s,l=(n=["a[href]","area[href]",'input:not([disabled]):not([type="hidden"]):not([aria-hidden])',"select:not([disabled]):not([aria-hidden])","textarea:not([disabled]):not([aria-hidden])","button:not([disabled]):not([aria-hidden])","iframe","object","embed","[contenteditable]",'[tabindex]:not([tabindex^="-"])'],i=function(){function o(e){var n=e.targetModal,i=e.triggers,a=void 0===i?[]:i,r=e.onShow,s=void 0===r?function(){}:r,l=e.onClose,c=void 0===l?function(){}:l,d=e.openTrigger,u=void 0===d?"data-micromodal-trigger":d,f=e.closeTrigger,h=void 0===f?"data-micromodal-close":f,v=e.openClass,g=void 0===v?"is-open":v,m=e.disableScroll,b=void 0!==m&&m,y=e.disableFocus,p=void 0!==y&&y,w=e.awaitCloseAnimation,E=void 0!==w&&w,k=e.awaitOpenAnimation,M=void 0!==k&&k,A=e.debugMode,C=void 0!==A&&A;!function(e,t){if(!(e instanceof t))throw new TypeError("Cannot call a class as a function")}(this,o),this.modal=document.getElementById(n),this.config={debugMode:C,disableScroll:b,openTrigger:u,closeTrigger:h,openClass:g,onShow:s,onClose:c,awaitCloseAnimation:E,awaitOpenAnimation:M,disableFocus:p},a.length>0&&this.registerTriggers.apply(this,t(a)),this.onClick=this.onClick.bind(this),this.onKeydown=this.onKeydown.bind(this)}var i,a,r;return i=o,(a=[{key:"registerTriggers",value:function(){for(var e=this,t=arguments.length,o=new Array(t),n=0;n<t;n++)o[n]=arguments[n];o.filter(Boolean).forEach((function(t){t.addEventListener("click",(function(t){return e.showModal(t)}))}))}},{key:"showModal",value:function(){var e=this,t=arguments.length>0&&void 0!==arguments[0]?arguments[0]:null;if(this.activeElement=document.activeElement,this.modal.setAttribute("aria-hidden","false"),this.modal.classList.add(this.config.openClass),this.scrollBehaviour("disable"),this.addEventListeners(),this.config.awaitOpenAnimation){var o=function t(){e.modal.removeEventListener("animationend",t,!1),e.setFocusToFirstNode()};this.modal.addEventListener("animationend",o,!1)}else this.setFocusToFirstNode();this.config.onShow(this.modal,this.activeElement,t)}},{key:"closeModal",value:function(){var e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:null,t=this.modal;if(this.modal.setAttribute("aria-hidden","true"),this.removeEventListeners(),this.scrollBehaviour("enable"),this.activeElement&&this.activeElement.focus&&this.activeElement.focus(),this.config.onClose(this.modal,this.activeElement,e),this.config.awaitCloseAnimation){var o=this.config.openClass;this.modal.addEventListener("animationend",(function e(){t.classList.remove(o),t.removeEventListener("animationend",e,!1)}),!1)}else t.classList.remove(this.config.openClass)}},{key:"closeModalById",value:function(e){this.modal=document.getElementById(e),this.modal&&this.closeModal()}},{key:"scrollBehaviour",value:function(e){if(this.config.disableScroll){var t=document.querySelector("body");switch(e){case"enable":Object.assign(t.style,{overflow:""});break;case"disable":Object.assign(t.style,{overflow:"hidden"})}}}},{key:"addEventListeners",value:function(){this.modal.addEventListener("touchstart",this.onClick),this.modal.addEventListener("click",this.onClick),document.addEventListener("keydown",this.onKeydown)}},{key:"removeEventListeners",value:function(){this.modal.removeEventListener("touchstart",this.onClick),this.modal.removeEventListener("click",this.onClick),document.removeEventListener("keydown",this.onKeydown)}},{key:"onClick",value:function(e){(e.target.hasAttribute(this.config.closeTrigger)||e.target.parentNode.hasAttribute(this.config.closeTrigger))&&(e.preventDefault(),e.stopPropagation(),this.closeModal(e))}},{key:"onKeydown",value:function(e){27===e.keyCode&&this.closeModal(e),9===e.keyCode&&this.retainFocus(e)}},{key:"getFocusableNodes",value:function(){var e=this.modal.querySelectorAll(n);return Array.apply(void 0,t(e))}},{key:"setFocusToFirstNode",value:function(){var e=this;if(!this.config.disableFocus){var t=this.getFocusableNodes();if(0!==t.length){var o=t.filter((function(t){return!t.hasAttribute(e.config.closeTrigger)}));o.length>0&&o[0].focus(),0===o.length&&t[0].focus()}}}},{key:"retainFocus",value:function(e){var t=this.getFocusableNodes();if(0!==t.length)if(t=t.filter((function(e){return null!==e.offsetParent})),this.modal.contains(document.activeElement)){var o=t.indexOf(document.activeElement);e.shiftKey&&0===o&&(t[t.length-1].focus(),e.preventDefault()),!e.shiftKey&&t.length>0&&o===t.length-1&&(t[0].focus(),e.preventDefault())}else t[0].focus()}}])&&e(i.prototype,a),r&&e(i,r),o}(),a=null,r=function(e){if(!document.getElementById(e))return console.warn("MicroModal: ❗Seems like you have missed %c'".concat(e,"'"),"background-color: #f8f9fa;color: #50596c;font-weight: bold;","ID somewhere in your code. Refer example below to resolve it."),console.warn("%cExample:","background-color: #f8f9fa;color: #50596c;font-weight: bold;",'<div class="modal" id="'.concat(e,'"></div>')),!1},s=function(e,t){if(function(e){e.length<=0&&(console.warn("MicroModal: ❗Please specify at least one %c'micromodal-trigger'","background-color: #f8f9fa;color: #50596c;font-weight: bold;","data attribute."),console.warn("%cExample:","background-color: #f8f9fa;color: #50596c;font-weight: bold;",'<a href="#" data-micromodal-trigger="my-modal"></a>'))}(e),!t)return!0;for(var o in t)r(o);return!0},{init:function(e){var o=Object.assign({},{openTrigger:"data-micromodal-trigger"},e),n=t(document.querySelectorAll("[".concat(o.openTrigger,"]"))),r=function(e,t){var o=[];return e.forEach((function(e){var n=e.attributes[t].value;void 0===o[n]&&(o[n]=[]),o[n].push(e)})),o}(n,o.openTrigger);if(!0!==o.debugMode||!1!==s(n,r))for(var l in r){var c=r[l];o.targetModal=l,o.triggers=t(c),a=new i(o)}},show:function(e,t){var o=t||{};o.targetModal=e,!0===o.debugMode&&!1===r(e)||(a&&a.removeEventListeners(),(a=new i(o)).showModal())},close:function(e){e?a.closeModalById(e):a.closeModal()}});return"undefined"!=typeof window&&(window.MicroModal=l),l}));  
  window.define = window.backupDefine; window.backupDefine = undefined;
  </script>
<script type="text/javascript">
  (function () {
    // forward keydown events so shortcuts can work in vscode, see:
    // https://github.com/microsoft/vscode/issues/65452#issuecomment-586485815
    if (window.parent.postMessage) {
      window.document.addEventListener('keydown', e => {
        const event = {
          type: "keydown",
          data: {
            altKey: e.altKey,
            code: e.code,
            ctrlKey: e.ctrlKey,
            isComposing: e.isComposing,
            key: e.key,
            location: e.location,
            metaKey: e.metaKey,
            repeat: e.repeat,
            shiftKey: e.shiftKey
          }
        };
        window.parent.postMessage(event, '*');
      });
    }
    // listen for execCommand messages
    window.addEventListener("message", function (event) {
      if (event.data.type === "devhost-exec-command") {
        window.document.execCommand(event.data.data);
      } 
    }, true);
    const logEntries = new Array(1000);
    logEntries.offset = 0;
    function recordLogEntry(entry) {
      logEntries[logEntries.offset++] = entry;
      logEntries.offset %= logEntries.length;
    }
    function getLogEntry(i) { // backwards, 0 is most recent
        return logEntries[(logEntries.offset - 1 - i + logEntries.length) % logEntries.length];
    }
    const protocol = window.location.protocol === "https:" ? "wss:" : "ws:";
    let path = window.location.pathname;
    if (!/\/$/.test(path)) path += "/";
    const devServerSocket = new WebSocket(protocol + "//" + window.location.host + path);
    let reloadPage = true;
    addEventListener("beforeunload", () => {
      // don't reload page in case of navigation events
      reloadPage = false;
    })
    devServerSocket.onopen = () => {
      console.log('Socket connection open. Listening for events.');
      // if the socket closes for any reason (e.g. this occurs in electron apps
      // when the computer suspends) then reload to reestablish the connection 
      devServerSocket.onclose = () => {
        if (reloadPage) {
          console.log('Socket connection closed. Reloading.');
          window.location.reload(true);
        }
      }
    };
    // append for errors that occur within the error window
    let lastError = 0;
    const kErrorWindow = 2500;
    function showError(msg) {
      const ansiUp = new AnsiUp
      const html = ansiUp.ansi_to_html(msg.trim());
      const display = document.getElementById("quarto-log-error-display");
      if (display.innerHTML && ((Date.now() - lastError) < kErrorWindow)) {
        display.innerHTML = display.innerHTML + "<br/>" +  html;
      } else {
        display.innerHTML = html
      }
      lastError = Date.now()
      MicroModal.show('quarto-log-error-modal', {
        awaitCloseAnimation: true,
        onClose: () => { display.innerHTML = ""; }
      });
      // post message to parent indicating we had an error
      if (window.parent.postMessage) {
        window.parent.postMessage({
          type: "error",
          msg: msg,
        }, "*");
      }
    }
    // see if there is already an error to show
    const renderError = document.getElementById("quarto-render-error");
    if (renderError) {
      showError(renderError.innerHTML.trim());
    }
    const normalizeTarget = (target) => {
      return target.replace(/\/index\.html/, "/")
    };
    devServerSocket.onmessage = (msg) => {
      if (msg.data.startsWith('reload')) {
        let target = normalizeTarget(msg.data.replace(/^reload/, ""));
        // prepend proxy path to target if we have one
        if (target) {
          const pathPrefix = 
            window.location.pathname.match(/^.*?\/p\/\w+\//) ||
            window.location.pathname.match(/^.*?\/user\/[\w\d]+\/proxy\/\d+\//);
          if (pathPrefix) {
            target = pathPrefix + target.slice(1);
          }
        }
        if (target && (target !== normalizeTarget(window.location.pathname))) {
          window.location.replace(target);
        } else {
          window.location.reload(true);
        }
      } else if (msg.data.startsWith('log:')) {
        const ansiUp = new AnsiUp
        const log = JSON.parse(msg.data.substr(4));
        recordLogEntry(log);
        if (log.levelName === "ERROR") {
          showError(log.msgFormatted)
        } else {
          // see if there is a knitr error to report
          const kExecutionHalted = "Execution halted";
          if (log.msg.indexOf(kExecutionHalted) !== -1) {
            // scan backwards for beginning of error
            const errorEntries = [getLogEntry(0).msgFormatted];
            for (let i=1; i<logEntries.length; i++) {
              const logEntry = getLogEntry(i);
              if (logEntry) {
                errorEntries.unshift(logEntry.msgFormatted);
                if (logEntry.msg.indexOf("Quitting from lines") !== -1) {
                  showError(errorEntries.join(""));
                  break;
                } else if (logEntry.msg.indexOf(kExecutionHalted) !== -1) {
                  break;
                }
              } else {
                break;
              }
            }
          }
        }
      }
    };
  })();
</script>

<script type="text/javascript">
  if (window.parent.postMessage) {
    // wait for message providing confirmation we are in a devhost
    window.addEventListener("message", function (event) {
      if (event.data.type === "devhost-init") {
        window.quartoDevhost = {
          openInputFile: function (line, column, highlight) {
            window.parent.postMessage({
              type: "openfile",
              file: "E:\workspace\course\2023Fall\2023Fall-CS-489-M01CS-509-M01-BIOINFORMATICSPROG\Assignment\hw2\HW.qmd",
              line: line,
              column: column,
              highlight: highlight
            }, event.origin);
          }
        };
      } else if (event.data.type === "goback") {
        window.history.back()
      } else if (event.data.type === "goforward") {
        window.history.forward()
      }
    }, true);
    // notify host of navigation (e.g. for 'pop out' command)
    window.parent.postMessage({
      type: "navigate",
      href: window.location.href,
      file: "E:\workspace\course\2023Fall\2023Fall-CS-489-M01CS-509-M01-BIOINFORMATICSPROG\Assignment\hw2\HW.qmd"
    }, "*");
  }
</script>
<extend-speak-to-chatgpt data-v-app=""><!----><div data-styled-id="rCS1w3zcxh" data-able-draggable="true" data-able-padding="true" data-able-clickable="true" class="moveable-control-box    moveable rCS1w3zcxh" style="position: absolute; display: none; visibility: visible; transform: translate3d(0px, 0px, 0px); --zoom: 1; --zoompx: 1px;"><div data-rotation="-1" data-line-key="render-line-0" data-direction="" class="moveable-line moveable-direction  " style="transform: translateY(-50%) translate(0px, 0px) rotate(0rad) scaleY(1); width: 0px;"></div><div data-rotation="-1" data-line-key="render-line-1" data-direction="" class="moveable-line moveable-direction  " style="transform: translateY(-50%) translate(0px, 0px) rotate(0rad) scaleY(1); width: 0px;"></div><div data-rotation="-1" data-line-key="render-line-2" data-direction="" class="moveable-line moveable-direction  " style="transform: translateY(-50%) translate(0px, 0px) rotate(0rad) scaleY(1); width: 0px;"></div><div data-rotation="-1" data-line-key="render-line-3" data-direction="" class="moveable-line moveable-direction  " style="transform: translateY(-50%) translate(0px, 0px) rotate(0rad) scaleY(1); width: 0px;"></div></div></extend-speak-to-chatgpt></body></html>