<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Zheng Cui">

<title>Hw3_Zheng</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="HW3_files/libs/clipboard/clipboard.min.js"></script>
<script src="HW3_files/libs/quarto-html/quarto.js"></script>
<script src="HW3_files/libs/quarto-html/popper.min.js"></script>
<script src="HW3_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="HW3_files/libs/quarto-html/anchor.min.js"></script>
<link href="HW3_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="HW3_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="HW3_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="HW3_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="HW3_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Hw3_Zheng</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Zheng Cui </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="quarto" class="level2">
<h2 class="anchored" data-anchor-id="quarto">Quarto</h2>
<p>Quarto enables you to weave together content and executable code into a finished document. To learn more about Quarto see <a href="https://quarto.org" class="uri">https://quarto.org</a>.</p>
</section>
<section id="running-code" class="level2">
<h2 class="anchored" data-anchor-id="running-code">Running Code</h2>
<p>When you click the <strong>Render</strong> button a document will be generated that includes both content and the output of embedded code. You can embed code like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> <span class="sc">+</span> <span class="dv">1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
</div>
<p>You can add options to executable code like this</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4</code></pre>
</div>
</div>
<p>The <code>echo: false</code> option disables the printing of code (only output is displayed).</p>
<section id="removing-the-version-number-from-an-ensembl-id" class="level3">
<h3 class="anchored" data-anchor-id="removing-the-version-number-from-an-ensembl-id">1. Removing the version number from an Ensembl ID</h3>
<section id="write-an-r-function-cut.version.numberensembl.id-using-regular-expression-to-extract-the-ensembl-id-without-the-attached-version-number.-for-examplegiven-the-ensembl-id-enst00000621592.8-of-the-myc-206-transcript-as-input-the-function-must-return-enst00000621592." class="level4">
<h4 class="anchored" data-anchor-id="write-an-r-function-cut.version.numberensembl.id-using-regular-expression-to-extract-the-ensembl-id-without-the-attached-version-number.-for-examplegiven-the-ensembl-id-enst00000621592.8-of-the-myc-206-transcript-as-input-the-function-must-return-enst00000621592.">(50%) Write an R function cut.version.number(ensembl.id) using regular expression to extract the Ensembl ID without the attached version number. For example,given the Ensembl ID “ENST00000621592.8” of the MYC-206 transcript as input, the function must return “ENST00000621592”.</h4>
</section>
<section id="the-function-should-be-vectorized-meaning-that-it-can-take-a-vector-of-ensembl-ids-as-input.-for-example-if-the-input-is-censt00000377970.6-enst00000259523.10-the-return-of-the-function-must-be-censt00000377970-enst00000259523.-it-should-avoid-any-for-loop-to-maintain-efficiency." class="level4">
<h4 class="anchored" data-anchor-id="the-function-should-be-vectorized-meaning-that-it-can-take-a-vector-of-ensembl-ids-as-input.-for-example-if-the-input-is-censt00000377970.6-enst00000259523.10-the-return-of-the-function-must-be-censt00000377970-enst00000259523.-it-should-avoid-any-for-loop-to-maintain-efficiency.">The function should be vectorized, meaning that it can take a vector of Ensembl IDs as input. For example, if the input is c(“ENST00000377970.6”, “ENST00000259523.10”), the return of the function must be c(“ENST00000377970”, “ENST00000259523”). It should avoid any for-loop to maintain efficiency.</h4>
<section id="removing-the-version-number-from-an-ensembl-id-1" class="level5">
<h5 class="anchored" data-anchor-id="removing-the-version-number-from-an-ensembl-id-1">Removing the version number from an Ensembl ID</h5>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>cut.version.number <span class="ot">&lt;-</span> <span class="cf">function</span>(ensembl.ids) {</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Use regular expression to remove the version number</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  remove_version <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">.</span><span class="sc">\\</span><span class="st">d+$"</span>, <span class="st">""</span>, ensembl.ids)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(remove_version)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Example usage:</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>ensembl_ids <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ENST00000621592.8"</span>, <span class="st">"ENST00000377970.6"</span>, <span class="st">"ENST00000259523.10"</span>)</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>remove_version <span class="ot">&lt;-</span> <span class="fu">cut.version.number</span>(ensembl_ids)</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(remove_version)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ENST00000621592" "ENST00000377970" "ENST00000259523"</code></pre>
</div>
</div>
</section>
<section id="apply-to-the-attributer-of-the-gff-file-filter-the-gene_id-out-and-then-remove-the-version-number-as-follows" class="level5">
<h5 class="anchored" data-anchor-id="apply-to-the-attributer-of-the-gff-file-filter-the-gene_id-out-and-then-remove-the-version-number-as-follows">Apply to the attributer of the GFF file, filter the gene_id out, and then remove the version number, as follows</h5>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co">#  import  GFF3 table</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>file <span class="ot">&lt;-</span> <span class="st">"E:/Language/R/gencode.v44.primary_assembly.annotation.gff3"</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>GFF <span class="ot">&lt;-</span> <span class="fu">read.table</span>(file, <span class="at">header =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(GFF) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"seqid"</span>, <span class="st">"source"</span>, <span class="st">"type"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>,</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"score"</span>, <span class="st">"strand"</span>, <span class="st">"phase"</span>, <span class="st">"attributes"</span>)</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(GFF) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(GFF)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>pattern <span class="ot">&lt;-</span> <span class="st">"gene_id=([^;]*)"</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="fu">regexec</span>(pattern, GFF[, <span class="st">"attributes"</span>])</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>gene_ID <span class="ot">&lt;-</span> <span class="fu">sapply</span>(</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">regmatches</span>(GFF[, <span class="st">"attributes"</span>], m), </span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(e) {<span class="fu">return</span>(e[<span class="dv">2</span>])}) </span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Add the extracted gene IDs to the GFF dataframe</span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>GFF<span class="sc">$</span>gene_ID <span class="ot">&lt;-</span> gene_ID</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Apply the cut.version.number function to the gene_id column of GFF</span></span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>GFF<span class="sc">$</span>trimmed_ids <span class="ot">&lt;-</span> <span class="fu">cut.version.number</span>(GFF<span class="sc">$</span>gene_ID)</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(GFF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  seqid source       type start   end score strand phase
1  chr1 HAVANA       gene 11869 14409     .      +     .
2  chr1 HAVANA transcript 11869 14409     .      +     .
3  chr1 HAVANA       exon 11869 12227     .      +     .
4  chr1 HAVANA       exon 12613 12721     .      +     .
5  chr1 HAVANA       exon 13221 14409     .      +     .
6  chr1 HAVANA       gene 12010 13670     .      +     .
                                                                                                                                                                                                                                                                                                                                          attributes
1                                                                                                                                                                                                                                  ID=ENSG00000290825.1;gene_id=ENSG00000290825.1;gene_type=lncRNA;gene_name=DDX11L2;level=2;tag=overlaps_pseudogene
2                                                ID=ENST00000456328.2;Parent=ENSG00000290825.1;gene_id=ENSG00000290825.1;transcript_id=ENST00000456328.2;gene_type=lncRNA;gene_name=DDX11L2;transcript_type=lncRNA;transcript_name=DDX11L2-202;level=2;transcript_support_level=1;tag=basic,Ensembl_canonical;havana_transcript=OTTHUMT00000362751.1
3 ID=exon:ENST00000456328.2:1;Parent=ENST00000456328.2;gene_id=ENSG00000290825.1;transcript_id=ENST00000456328.2;gene_type=lncRNA;gene_name=DDX11L2;transcript_type=lncRNA;transcript_name=DDX11L2-202;exon_number=1;exon_id=ENSE00002234944.1;level=2;transcript_support_level=1;tag=basic,Ensembl_canonical;havana_transcript=OTTHUMT00000362751.1
4 ID=exon:ENST00000456328.2:2;Parent=ENST00000456328.2;gene_id=ENSG00000290825.1;transcript_id=ENST00000456328.2;gene_type=lncRNA;gene_name=DDX11L2;transcript_type=lncRNA;transcript_name=DDX11L2-202;exon_number=2;exon_id=ENSE00003582793.1;level=2;transcript_support_level=1;tag=basic,Ensembl_canonical;havana_transcript=OTTHUMT00000362751.1
5 ID=exon:ENST00000456328.2:3;Parent=ENST00000456328.2;gene_id=ENSG00000290825.1;transcript_id=ENST00000456328.2;gene_type=lncRNA;gene_name=DDX11L2;transcript_type=lncRNA;transcript_name=DDX11L2-202;exon_number=3;exon_id=ENSE00002312635.1;level=2;transcript_support_level=1;tag=basic,Ensembl_canonical;havana_transcript=OTTHUMT00000362751.1
6                                                                                                                                                                          ID=ENSG00000223972.6;gene_id=ENSG00000223972.6;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;level=2;hgnc_id=HGNC:37102;havana_gene=OTTHUMG00000000961.2
            gene_ID     trimmed_ids
1 ENSG00000290825.1 ENSG00000290825
2 ENSG00000290825.1 ENSG00000290825
3 ENSG00000290825.1 ENSG00000290825
4 ENSG00000290825.1 ENSG00000290825
5 ENSG00000290825.1 ENSG00000290825
6 ENSG00000223972.6 ENSG00000223972</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="reading-genome-annotation-in-gtf-format" class="level3">
<h3 class="anchored" data-anchor-id="reading-genome-annotation-in-gtf-format">2. Reading genome annotation in GTF format</h3>
<section id="the-gtf-format-is-as-popular-as-the-gff3-format.-please-describe-three-differences-between-the-two-formats." class="level4">
<h4 class="anchored" data-anchor-id="the-gtf-format-is-as-popular-as-the-gff3-format.-please-describe-three-differences-between-the-two-formats.">(15%) 2.1 The GTF format is as popular as the GFF3 format. Please describe three differences between the two formats.</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co">#  import  GFF3 table</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>file <span class="ot">&lt;-</span> <span class="st">"E:/Language/R/gencode.v44.annotation.gtf"</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>GTF <span class="ot">&lt;-</span> <span class="fu">read.table</span>(file, <span class="at">header =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(GTF) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"seqid"</span>, <span class="st">"source"</span>, <span class="st">"type"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>,</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"score"</span>, <span class="st">"strand"</span>, <span class="st">"phase"</span>, <span class="st">"attributes"</span>)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(GTF) <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(GTF)</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(GTF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  seqid source       type start   end score strand phase
1  chr1 HAVANA       gene 11869 14409     .      +     .
2  chr1 HAVANA transcript 11869 14409     .      +     .
3  chr1 HAVANA       exon 11869 12227     .      +     .
4  chr1 HAVANA       exon 12613 12721     .      +     .
5  chr1 HAVANA       exon 13221 14409     .      +     .
6  chr1 HAVANA       gene 12010 13670     .      +     .
                                                                                                                                                                                                                                                                                                      attributes
1                                                                                                                                                                                                              gene_id ENSG00000290825.1; gene_type lncRNA; gene_name DDX11L2; level 2; tag overlaps_pseudogene;
2                                           gene_id ENSG00000290825.1; transcript_id ENST00000456328.2; gene_type lncRNA; gene_name DDX11L2; transcript_type lncRNA; transcript_name DDX11L2-202; level 2; transcript_support_level 1; tag basic; tag Ensembl_canonical; havana_transcript OTTHUMT00000362751.1;
3 gene_id ENSG00000290825.1; transcript_id ENST00000456328.2; gene_type lncRNA; gene_name DDX11L2; transcript_type lncRNA; transcript_name DDX11L2-202; exon_number 1; exon_id ENSE00002234944.1; level 2; transcript_support_level 1; tag basic; tag Ensembl_canonical; havana_transcript OTTHUMT00000362751.1;
4 gene_id ENSG00000290825.1; transcript_id ENST00000456328.2; gene_type lncRNA; gene_name DDX11L2; transcript_type lncRNA; transcript_name DDX11L2-202; exon_number 2; exon_id ENSE00003582793.1; level 2; transcript_support_level 1; tag basic; tag Ensembl_canonical; havana_transcript OTTHUMT00000362751.1;
5 gene_id ENSG00000290825.1; transcript_id ENST00000456328.2; gene_type lncRNA; gene_name DDX11L2; transcript_type lncRNA; transcript_name DDX11L2-202; exon_number 3; exon_id ENSE00002312635.1; level 2; transcript_support_level 1; tag basic; tag Ensembl_canonical; havana_transcript OTTHUMT00000362751.1;
6                                                                                                                                                     gene_id ENSG00000223972.6; gene_type transcribed_unprocessed_pseudogene; gene_name DDX11L1; level 2; hgnc_id HGNC:37102; havana_gene OTTHUMG00000000961.2;</code></pre>
</div>
</div>
</section>
<section id="answer" class="level4">
<h4 class="anchored" data-anchor-id="answer">Answer：</h4>
<p>After importing, we can see: 1, mainly the difference between attributes, GTF uses Spaces, while GFF3 equals the sign =, the attribute format in GFF3: ** Although GFF3 uses key-value pairs in attributes, the keys and values themselves can sometimes have additional structure. For example, some properties may include multiple values separated by commas, or they may use different key-value pairs to convey additional information. 2. GFF3 has an ID and parent ID in its properties, but GTF does not 3. GTF usually contains the required basic attributes (e.g., gene_id, transcript_id) as mandatory attributes, while GFF3 allows greater flexibility in attribute names. 4. Optional fields: GFF3 contains optional fields for additional information, while GTF is more concise and may lack some of these optional details. 5, the file size is roughly the same, the compressed file size is also roughly the same 6, the compressed file size is also roughly the same.</p>
</section>
<section id="use-regular-expression-to-extract-gene-names-for-every-feature-in-the-file.-report-the-runtime-of-your-code-using-the-system.time-function." class="level4">
<h4 class="anchored" data-anchor-id="use-regular-expression-to-extract-gene-names-for-every-feature-in-the-file.-report-the-runtime-of-your-code-using-the-system.time-function.">(35%) 2.2 Use regular expression to extract gene names for every feature in the file. Report the runtime of your code using the system.time() function.</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>time.vec <span class="ot">&lt;-</span> <span class="fu">system.time</span>({</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>pattern <span class="ot">&lt;-</span> <span class="st">"gene_name ([^;]+);"</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a> m <span class="ot">&lt;-</span> <span class="fu">regexec</span>(pattern, GTF[, <span class="st">"attributes"</span>])</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  gene_names <span class="ot">&lt;-</span> <span class="fu">sapply</span>(</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">regmatches</span>(GTF[, <span class="st">"attributes"</span>], m), </span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">function</span>(e) {<span class="fu">return</span>(e[<span class="dv">2</span>])}) </span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Add the extracted gene names as a new column</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>GTF<span class="sc">$</span>gene_name <span class="ot">&lt;-</span> gene_names</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>  <span class="co">#</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Very slow alternative to extract gene names:    </span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>  <span class="co">#   d &lt;- data.frame(regmatches(gff3[, "tag"], m))    </span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>  <span class="co">#   gene_names &lt;- d[2, ]  </span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>  } )</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(time.vec[<span class="st">"user.self"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>user.self 
     4.81 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(GTF<span class="sc">$</span>gene_name)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "DDX11L2" "DDX11L2" "DDX11L2" "DDX11L2" "DDX11L2" "DDX11L1"</code></pre>
</div>
</div>
</section>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>